// avbridge/dist/chunk-OFJYEITB.cjs
'use strict';

var chunkWRKO6Q42_cjs = require('./chunk-WRKO6Q42.cjs');
var chunkVLI3Y6IJ_cjs = require('./chunk-VLI3Y6IJ.cjs');
var chunk2IJ66NTD_cjs = require('./chunk-2IJ66NTD.cjs');
var chunkHZUVMXBN_cjs = require('./chunk-HZUVMXBN.cjs');
var chunkG4APZMCP_cjs = require('./chunk-G4APZMCP.cjs');
var chunkF3LQJKXK_cjs = require('./chunk-F3LQJKXK.cjs');

// src/util/codec-strings.ts
function videoCodecString(track) {
  if (track.codecString) return track.codecString;
  switch (track.codec) {
    case "h264": {
      const profileHex = profileToHex(track.profile) ?? "64";
      const constraint = "00";
      const level = ((track.level ?? 40) & 255).toString(16).padStart(2, "0");
      return `avc1.${profileHex}${constraint}${level}`;
    }
    case "h265":
      return "hvc1.1.6.L93.B0";
    case "vp8":
      return "vp8";
    case "vp9":
      return "vp09.00.10.08";
    case "av1":
      return "av01.0.04M.08";
    default:
      return null;
  }
}
function profileToHex(profile) {
  if (!profile) return null;
  const p = profile.toLowerCase();
  if (p.includes("baseline")) return "42";
  if (p.includes("main")) return "4d";
  if (p.includes("high 10")) return "6e";
  if (p.includes("high 4:2:2")) return "7a";
  if (p.includes("high 4:4:4")) return "f4";
  if (p.includes("high")) return "64";
  return null;
}
function audioCodecString(track) {
  if (track.codecString) return track.codecString;
  switch (track.codec) {
    case "aac":
      return "mp4a.40.2";
    // AAC-LC
    case "mp3":
      return "mp4a.40.34";
    case "opus":
      return "opus";
    case "vorbis":
      return "vorbis";
    case "flac":
      return "flac";
    default:
      return null;
  }
}
function mp4MimeFor(video, audio) {
  const v = videoCodecString(video);
  if (!v) return null;
  const codecs = audio ? `${v},${audioCodecString(audio) ?? ""}`.replace(/,$/, "") : v;
  return `video/mp4; codecs="${codecs}"`;
}
function mseSupports(mime) {
  if (typeof MediaSource === "undefined") return false;
  try {
    return MediaSource.isTypeSupported(mime);
  } catch {
    return false;
  }
}

// src/classify/rules.ts
var NATIVE_VIDEO_CODECS = /* @__PURE__ */ new Set(["h264", "h265", "vp8", "vp9", "av1"]);
var NATIVE_AUDIO_CODECS = /* @__PURE__ */ new Set([
  "aac",
  "mp3",
  "opus",
  "vorbis",
  "flac"
]);
var FALLBACK_VIDEO_CODECS = /* @__PURE__ */ new Set([
  "wmv3",
  "vc1",
  "mpeg4",
  "rv10",
  "rv20",
  "rv30",
  "rv40",
  "mpeg2",
  "mpeg1",
  "theora",
  "dv",
  "hq_hqa",
  "rawvideo",
  "qtrle",
  "png",
  "vp6f"
]);
var FALLBACK_AUDIO_CODECS = /* @__PURE__ */ new Set([
  "wmav2",
  "wmapro",
  "ac3",
  "eac3",
  "cook",
  "ra_144",
  "ra_288",
  "sipr",
  "atrac3",
  "dts",
  "truehd"
]);
var NATIVE_CONTAINERS = /* @__PURE__ */ new Set([
  "mp4",
  "mov",
  "webm",
  "ogg",
  "wav",
  "mp3",
  "flac",
  "adts"
]);
var REMUXABLE_CONTAINERS = /* @__PURE__ */ new Set([
  "mp4",
  "mov",
  "mkv",
  "webm",
  "ogg",
  "wav",
  "mp3",
  "flac",
  "adts",
  "mpegts"
]);
function classifyContext(ctx) {
  const video = ctx.videoTracks[0];
  const audio = ctx.audioTracks[0];
  if (!video) {
    if (NATIVE_CONTAINERS.has(ctx.container) && (!audio || NATIVE_AUDIO_CODECS.has(audio.codec))) {
      return {
        class: "NATIVE",
        strategy: "native",
        reason: `audio-only ${ctx.container} with native codec`
      };
    }
    if (audio && FALLBACK_AUDIO_CODECS.has(audio.codec)) {
      return {
        class: "FALLBACK_REQUIRED",
        strategy: "fallback",
        reason: `audio codec "${audio.codec}" requires WASM decode`
      };
    }
    if (REMUXABLE_CONTAINERS.has(ctx.container)) {
      return {
        class: "REMUX_CANDIDATE",
        strategy: "remux",
        reason: `audio-only file in non-native container "${ctx.container}"`
      };
    }
    return {
      class: "FALLBACK_REQUIRED",
      strategy: "fallback",
      reason: `audio-only file in "${ctx.container}" (not remuxable by mediabunny)`
    };
  }
  if (FALLBACK_VIDEO_CODECS.has(video.codec)) {
    return {
      class: "FALLBACK_REQUIRED",
      strategy: "fallback",
      reason: `video codec "${video.codec}" has no browser decoder; WASM fallback required`
    };
  }
  const audioNeedsFallback = audio && (FALLBACK_AUDIO_CODECS.has(audio.codec) || !NATIVE_AUDIO_CODECS.has(audio.codec));
  if (audioNeedsFallback) {
    if (NATIVE_VIDEO_CODECS.has(video.codec) && webCodecsAvailable()) {
      return {
        class: "HYBRID_CANDIDATE",
        strategy: "hybrid",
        reason: `video "${video.codec}" is hardware-decodable via WebCodecs; audio "${audio.codec}" decoded in software by libav`,
        fallbackChain: ["fallback"]
      };
    }
    return {
      class: "FALLBACK_REQUIRED",
      strategy: "fallback",
      reason: `audio codec "${audio.codec}" has no browser decoder; WASM fallback required`
    };
  }
  if (!NATIVE_VIDEO_CODECS.has(video.codec)) {
    return {
      class: "FALLBACK_REQUIRED",
      strategy: "fallback",
      reason: `unknown video codec "${video.codec}", routing to fallback`
    };
  }
  const isNativeContainer = NATIVE_CONTAINERS.has(ctx.container);
  if (isNativeContainer && isSafeNativeCombo(video, audio)) {
    const mime = mp4MimeFor(video, audio);
    if (mime && mseSupports(mime)) {
      return {
        class: "NATIVE",
        strategy: "native",
        reason: `${ctx.container} + ${video.codec}${audio ? "/" + audio.codec : ""} plays natively`
      };
    }
    if (mime == null || typeof MediaSource === "undefined") {
      return {
        class: "NATIVE",
        strategy: "native",
        reason: `${ctx.container} + ${video.codec}${audio ? "/" + audio.codec : ""} (heuristic native)`
      };
    }
  }
  if (isNativeContainer && isRiskyNative(video)) {
    return {
      class: "RISKY_NATIVE",
      strategy: "native",
      reason: `${video.codec} ${video.profile ?? ""} ${video.bitDepth ?? 8}-bit may stutter on mobile; will escalate to remux on stall`,
      fallbackChain: ["remux", "hybrid", "fallback"]
    };
  }
  if (REMUXABLE_CONTAINERS.has(ctx.container)) {
    const mime = mp4MimeFor(video, audio);
    if (mime && typeof MediaSource !== "undefined" && !mseSupports(mime)) {
      if (webCodecsAvailable()) {
        return {
          class: "HYBRID_CANDIDATE",
          strategy: "hybrid",
          reason: `${ctx.container} container with ${video.codec}${audio ? "/" + audio.codec : ""}; MSE rejects the remux target mime \u2014 routing to WebCodecs hardware decode`,
          fallbackChain: ["fallback"]
        };
      }
      return {
        class: "FALLBACK_REQUIRED",
        strategy: "fallback",
        reason: `${ctx.container} container with ${video.codec}${audio ? "/" + audio.codec : ""}; MSE rejects the remux target mime and WebCodecs is unavailable \u2014 falling back to WASM decode`
      };
    }
    const fallbackChain = webCodecsAvailable() ? ["hybrid", "fallback"] : ["fallback"];
    return {
      class: "REMUX_CANDIDATE",
      strategy: "remux",
      reason: `${ctx.container} container with native-supported codecs \u2014 remux to fragmented MP4 for reliable playback`,
      fallbackChain
    };
  }
  if (webCodecsAvailable()) {
    return {
      class: "HYBRID_CANDIDATE",
      strategy: "hybrid",
      reason: `${ctx.container} container requires libav demux; codecs (${video.codec}${audio ? "/" + audio.codec : ""}) are hardware-decodable via WebCodecs`,
      fallbackChain: ["fallback"]
    };
  }
  return {
    class: "FALLBACK_REQUIRED",
    strategy: "fallback",
    reason: `${ctx.container} container cannot be remuxed by mediabunny; falling back to WASM decode (${video.codec}${audio ? "/" + audio.codec : ""})`
  };
}
function webCodecsAvailable() {
  return typeof globalThis.VideoDecoder !== "undefined";
}
function isSafeNativeCombo(video, audio) {
  if (video.codec === "h264") {
    if (video.bitDepth && video.bitDepth > 8) return false;
    if (video.pixelFormat && !/yuv420p$/.test(video.pixelFormat)) return false;
  }
  if (audio && !NATIVE_AUDIO_CODECS.has(audio.codec)) return false;
  return true;
}
function isRiskyNative(video) {
  if (video.bitDepth && video.bitDepth > 8) return true;
  if (video.pixelFormat && /yuv4(2[24]|44)/.test(video.pixelFormat)) return true;
  if (video.width > 3840 || video.height > 2160) return true;
  if (video.fps && video.fps > 60) return true;
  return false;
}

// src/events.ts
var TypedEmitter = class {
  listeners = {};
  sticky = {};
  on(event, fn) {
    let set = this.listeners[event];
    if (!set) {
      set = /* @__PURE__ */ new Set();
      this.listeners[event] = set;
    }
    set.add(fn);
    if (Object.prototype.hasOwnProperty.call(this.sticky, event)) {
      try {
        fn(this.sticky[event]);
      } catch (err) {
        console.error("[avbridge] listener threw replaying sticky value:", err);
      }
    }
    return () => this.off(event, fn);
  }
  off(event, fn) {
    this.listeners[event]?.delete(fn);
  }
  emit(event, payload) {
    const set = this.listeners[event];
    if (!set) return;
    for (const fn of [...set]) {
      try {
        fn(payload);
      } catch (err) {
        console.error("[avbridge] listener threw:", err);
      }
    }
  }
  /**
   * Like {@link emit} but also remembers the value so future subscribers
   * receive it on `on()`. Use for one-shot state-snapshot events.
   */
  emitSticky(event, payload) {
    this.sticky[event] = payload;
    this.emit(event, payload);
  }
  removeAll() {
    this.listeners = {};
    this.sticky = {};
  }
};

// src/diagnostics.ts
var Diagnostics = class {
  container = "unknown";
  videoCodec;
  audioCodec;
  width;
  height;
  fps;
  duration;
  strategy = "pending";
  strategyClass = "pending";
  reason = "";
  probedBy;
  sourceType;
  transport;
  rangeSupported;
  runtime = {};
  lastError;
  strategyHistory = [];
  recordProbe(ctx) {
    this.container = ctx.container;
    this.probedBy = ctx.probedBy;
    this.duration = ctx.duration;
    const v = ctx.videoTracks[0];
    if (v) {
      this.videoCodec = v.codec;
      this.width = v.width;
      this.height = v.height;
      this.fps = v.fps;
    }
    const a = ctx.audioTracks[0];
    if (a) this.audioCodec = a.codec;
    const src = ctx.source;
    if (typeof src === "string" || src instanceof URL) {
      this.sourceType = "url";
      this.transport = "http-range";
      this.rangeSupported = void 0;
    } else {
      this.sourceType = "blob";
      this.transport = "memory";
      this.rangeSupported = false;
    }
  }
  /**
   * Called by a strategy once it has a confirmed answer about how the
   * source bytes are actually flowing (e.g. after the libav HTTP block
   * reader's initial Range probe succeeded). Lets diagnostics report the
   * truth instead of an input-type heuristic.
   */
  recordTransport(transport, rangeSupported) {
    this.transport = transport;
    this.rangeSupported = rangeSupported;
  }
  recordClassification(c) {
    this.strategy = c.strategy;
    this.strategyClass = c.class;
    this.reason = c.reason;
  }
  recordRuntime(stats) {
    const {
      _transport,
      _rangeSupported,
      ...rest
    } = stats;
    if (_transport != null && typeof _rangeSupported === "boolean") {
      this.recordTransport(_transport, _rangeSupported);
    }
    this.runtime = { ...this.runtime, ...rest };
  }
  recordStrategySwitch(strategy, reason) {
    this.strategy = strategy;
    this.reason = reason;
    this.strategyHistory.push({ strategy, reason, at: Date.now() });
  }
  recordError(err) {
    this.lastError = err;
  }
  snapshot() {
    const snap = {
      container: this.container,
      videoCodec: this.videoCodec,
      audioCodec: this.audioCodec,
      width: this.width,
      height: this.height,
      fps: this.fps,
      duration: this.duration,
      strategy: this.strategy,
      strategyClass: this.strategyClass,
      reason: this.reason,
      probedBy: this.probedBy,
      sourceType: this.sourceType,
      transport: this.transport,
      rangeSupported: this.rangeSupported,
      runtime: { ...this.runtime, ...this.lastError ? { error: this.lastError.message } : {} },
      strategyHistory: this.strategyHistory.length > 0 ? [...this.strategyHistory] : void 0
    };
    return Object.freeze(snap);
  }
};

// src/plugins/registry.ts
var PluginRegistry = class {
  plugins = [];
  register(plugin, prepend = false) {
    if (prepend) this.plugins.unshift(plugin);
    else this.plugins.push(plugin);
  }
  all() {
    return this.plugins;
  }
  /**
   * Find the first plugin that claims this context AND its name matches the
   * strategy. Built-in strategy plugins are named exactly `"native"`,
   * `"remux"`, `"fallback"`.
   */
  findFor(context, strategy) {
    for (const p of this.plugins) {
      if (p.name === strategy && p.canHandle(context)) return p;
    }
    return null;
  }
};

// src/strategies/native.ts
async function createNativeSession(context, video) {
  const { url, revoke } = sourceToVideoUrl(context.source);
  video.src = url;
  await new Promise((resolve, reject) => {
    const onMeta = () => {
      cleanup();
      resolve();
    };
    const onError = () => {
      cleanup();
      reject(new Error(`<video> failed to load: ${video.error?.message ?? "unknown"}`));
    };
    const cleanup = () => {
      video.removeEventListener("loadedmetadata", onMeta);
      video.removeEventListener("error", onError);
    };
    video.addEventListener("loadedmetadata", onMeta);
    video.addEventListener("error", onError);
  });
  let stats = { framesDecoded: 0, framesDropped: 0 };
  return {
    strategy: "native",
    async play() {
      await video.play();
    },
    pause() {
      video.pause();
    },
    async seek(time) {
      video.currentTime = time;
    },
    async setAudioTrack(id) {
      const tracks = video.audioTracks;
      if (!tracks || tracks.length === 0) return;
      for (let i = 0; i < tracks.length; i++) {
        tracks[i].enabled = tracks[i].id === String(id) || i === id;
      }
    },
    async setSubtitleTrack(id) {
      const tracks = video.textTracks;
      for (let i = 0; i < tracks.length; i++) {
        tracks[i].mode = i === id ? "showing" : "disabled";
      }
    },
    async destroy() {
      video.pause();
      video.removeAttribute("src");
      video.load();
      revoke?.();
    },
    getCurrentTime() {
      return video.currentTime || 0;
    },
    getRuntimeStats() {
      const q = video.getVideoPlaybackQuality?.();
      if (q) {
        stats = {
          framesDecoded: q.totalVideoFrames,
          framesDropped: q.droppedVideoFrames
        };
      }
      return { ...stats, decoderType: "native" };
    }
  };
}
function sourceToVideoUrl(source) {
  if (source instanceof Blob) {
    const url = URL.createObjectURL(source);
    return { url, revoke: () => URL.revokeObjectURL(url) };
  }
  if (source instanceof ArrayBuffer || source instanceof Uint8Array) {
    const blob = new Blob([source]);
    const url = URL.createObjectURL(blob);
    return { url, revoke: () => URL.revokeObjectURL(url) };
  }
  if (typeof source === "string") return { url: source };
  if (source instanceof URL) return { url: source.toString() };
  throw new TypeError("native strategy: unsupported source type");
}

// src/strategies/remux/mse.ts
var MseSink = class {
  constructor(options) {
    this.options = options;
    if (typeof MediaSource === "undefined") {
      throw new chunk2IJ66NTD_cjs.AvbridgeError(
        chunk2IJ66NTD_cjs.ERR_MSE_NOT_SUPPORTED,
        "MediaSource Extensions (MSE) are not supported in this environment.",
        "MSE is required for the remux strategy. Use a browser that supports MSE, or try the fallback strategy."
      );
    }
    if (!MediaSource.isTypeSupported(options.mime)) {
      throw new chunk2IJ66NTD_cjs.AvbridgeError(
        chunk2IJ66NTD_cjs.ERR_MSE_CODEC_NOT_SUPPORTED,
        `This browser's MSE does not support "${options.mime}".`,
        "The codec combination can't be played via remux in this browser. The player will try the next strategy automatically."
      );
    }
    this.mediaSource = new MediaSource();
    this.objectUrl = URL.createObjectURL(this.mediaSource);
    options.video.src = this.objectUrl;
    this.readyPromise = new Promise((resolve, reject) => {
      this.resolveReady = resolve;
      this.rejectReady = reject;
    });
    this.mediaSource.addEventListener("sourceopen", () => {
      try {
        this.sourceBuffer = this.mediaSource.addSourceBuffer(options.mime);
        this.sourceBuffer.mode = "segments";
        this.sourceBuffer.addEventListener("updateend", () => this.pump());
        this.resolveReady();
        options.onReady?.();
      } catch (err) {
        this.rejectReady(err instanceof Error ? err : new Error(String(err)));
      }
    });
  }
  options;
  mediaSource;
  sourceBuffer = null;
  queue = [];
  endOfStreamCalled = false;
  destroyed = false;
  readyPromise;
  resolveReady;
  rejectReady;
  objectUrl;
  ready() {
    return this.readyPromise;
  }
  /** Queue a chunk of fMP4 bytes (init segment or media segment). */
  append(chunk) {
    if (this.destroyed) return;
    const ab = chunk instanceof Uint8Array ? chunk.buffer.slice(chunk.byteOffset, chunk.byteOffset + chunk.byteLength) : chunk;
    this.queue.push(ab);
    this.pump();
  }
  pump() {
    const sb = this.sourceBuffer;
    if (!sb || sb.updating) return;
    if (sb.buffered.length > 0) {
      if (this.pendingSeekTime !== null) {
        this.options.video.currentTime = this.pendingSeekTime;
        this.pendingSeekTime = null;
      } else if (!this.hasSnappedToFirstBuffered) {
        const v = this.options.video;
        const firstStart = sb.buffered.start(0);
        const firstEnd = sb.buffered.end(0);
        if (v.currentTime < firstStart || v.currentTime > firstEnd) {
          v.currentTime = firstStart;
        }
        this.hasSnappedToFirstBuffered = true;
      }
      if (this.playOnSeek) {
        this.playOnSeek = false;
        this.options.video.play().catch(() => {
        });
      }
    }
    const next = this.queue.shift();
    if (!next) return;
    try {
      sb.appendBuffer(next);
    } catch (err) {
      if (err.name === "QuotaExceededError") {
        this.evict();
        try {
          sb.appendBuffer(next);
          return;
        } catch {
        }
      }
      this.rejectReady(err instanceof Error ? err : new Error(String(err)));
    }
  }
  evict() {
    const sb = this.sourceBuffer;
    if (!sb || sb.buffered.length === 0) return;
    const start = sb.buffered.start(0);
    const current = this.options.video.currentTime;
    if (current - start > 10) {
      try {
        sb.remove(start, current - 10);
      } catch {
      }
    }
  }
  /** Indicate the source is finished. Future seeks past the end will fail. */
  endOfStream() {
    if (this.endOfStreamCalled || this.destroyed) return;
    this.endOfStreamCalled = true;
    const tryEnd = () => {
      if (this.queue.length > 0 || this.sourceBuffer?.updating) {
        this.sourceBuffer?.addEventListener("updateend", tryEnd, { once: true });
        return;
      }
      try {
        if (this.mediaSource.readyState === "open") {
          this.mediaSource.endOfStream();
        }
      } catch {
      }
    };
    tryEnd();
  }
  /** Seconds of media buffered ahead of the current playback position. */
  bufferedAhead() {
    const sb = this.sourceBuffer;
    if (!sb || sb.buffered.length === 0) return 0;
    const current = this.options.video.currentTime;
    for (let i = 0; i < sb.buffered.length; i++) {
      if (sb.buffered.start(i) <= current && sb.buffered.end(i) > current) {
        return sb.buffered.end(i) - current;
      }
    }
    return 0;
  }
  /** Total seconds of media buffered across all ranges. */
  totalBuffered() {
    const sb = this.sourceBuffer;
    if (!sb || sb.buffered.length === 0) return 0;
    let total = 0;
    for (let i = 0; i < sb.buffered.length; i++) {
      total += sb.buffered.end(i) - sb.buffered.start(i);
    }
    return total;
  }
  /** Number of chunks waiting in the append queue. */
  queueLength() {
    return this.queue.length;
  }
  /** Time to seek to once the SourceBuffer has data at this position. */
  pendingSeekTime = null;
  /** Whether to resume playback after the deferred seek completes. */
  playOnSeek = false;
  /**
   * On the very first data arrival, if `currentTime` falls outside the first
   * buffered range, snap it to the start of that range. MPEG-TS sources
   * commonly start their PTS at a non-zero value (e.g. ~1.5s); without this
   * snap, the video element sits at `currentTime=0` waiting forever for
   * data that doesn't exist.
   */
  hasSnappedToFirstBuffered = false;
  /** Request that playback resumes automatically once the deferred seek fires. */
  setPlayOnSeek(play) {
    this.playOnSeek = play;
  }
  /**
   * Discard all buffered media and schedule a deferred seek. The actual
   * `video.currentTime` assignment happens in `pump()` once the SourceBuffer
   * has data at the target position — setting it earlier causes the browser
   * to snap back to the nearest buffered range.
   */
  invalidate(seekTime) {
    const sb = this.sourceBuffer;
    this.queue = [];
    this.pendingSeekTime = seekTime;
    this.hasSnappedToFirstBuffered = true;
    if (!sb || sb.buffered.length === 0) return;
    try {
      const start = sb.buffered.start(0);
      const end = sb.buffered.end(sb.buffered.length - 1);
      sb.remove(start, end);
    } catch {
    }
  }
  destroy() {
    this.destroyed = true;
    this.queue = [];
    try {
      if (this.mediaSource.readyState === "open") this.mediaSource.endOfStream();
    } catch {
    }
    URL.revokeObjectURL(this.objectUrl);
  }
};

// src/strategies/remux/pipeline.ts
async function createRemuxPipeline(ctx, video) {
  const mb = await import('mediabunny');
  const videoTrackInfo = ctx.videoTracks[0];
  if (!videoTrackInfo) throw new Error("remux: source has no video track");
  const mbVideoCodec = chunkVLI3Y6IJ_cjs.avbridgeVideoToMediabunny(videoTrackInfo.codec);
  if (!mbVideoCodec) {
    throw new Error(`remux: video codec "${videoTrackInfo.codec}" is not supported by mediabunny output`);
  }
  const input = new mb.Input({
    source: await chunkVLI3Y6IJ_cjs.buildMediabunnySourceFromInput(mb, ctx.source),
    formats: mb.ALL_FORMATS
  });
  const allTracks = await input.getTracks();
  const inputVideo = allTracks.find((t) => t.id === videoTrackInfo.id && t.isVideoTrack());
  if (!inputVideo || !inputVideo.isVideoTrack()) {
    throw new Error("remux: video track not found in input");
  }
  const videoConfig = await inputVideo.getDecoderConfig();
  const videoSink = new mb.EncodedPacketSink(inputVideo);
  let selectedAudioTrackId = ctx.audioTracks[0]?.id ?? null;
  let inputAudio = null;
  let mbAudioCodec = null;
  let audioSink = null;
  let audioConfig = null;
  async function rebuildAudio() {
    if (selectedAudioTrackId == null) {
      inputAudio = null;
      mbAudioCodec = null;
      audioSink = null;
      audioConfig = null;
      return;
    }
    const trackInfo = ctx.audioTracks.find((t) => t.id === selectedAudioTrackId);
    if (!trackInfo) {
      throw new Error(`remux: no audio track with id ${selectedAudioTrackId}`);
    }
    const newInput = allTracks.find((t) => t.id === trackInfo.id && t.isAudioTrack());
    if (!newInput || !newInput.isAudioTrack()) {
      throw new Error("remux: audio track not found in input");
    }
    inputAudio = newInput;
    mbAudioCodec = chunkVLI3Y6IJ_cjs.avbridgeAudioToMediabunny(trackInfo.codec);
    audioSink = new mb.EncodedPacketSink(newInput);
    audioConfig = await newInput.getDecoderConfig();
  }
  await rebuildAudio();
  let sink = null;
  const stats = { videoPackets: 0, audioPackets: 0, bytesWritten: 0, fragments: 0 };
  let destroyed = false;
  let pumpToken = 0;
  let pendingAutoPlay = false;
  let pendingStartTime = 0;
  let currentOutput = null;
  function createOutput() {
    if (currentOutput) {
      try {
        void currentOutput.cancel();
      } catch {
      }
    }
    let mimePromise = null;
    const myToken = pumpToken;
    const writable = new WritableStream({
      write: async (chunk) => {
        if (destroyed || pumpToken !== myToken) return;
        if (!sink) {
          const mime = await (mimePromise ??= output.getMimeType());
          if (destroyed || pumpToken !== myToken) return;
          sink = new MseSink({ mime, video });
          await sink.ready();
          if (destroyed || pumpToken !== myToken) return;
          if (pendingStartTime > 0) {
            sink.invalidate(pendingStartTime);
          }
          sink.setPlayOnSeek(pendingAutoPlay);
        }
        while (sink && !destroyed && pumpToken === myToken && (sink.queueLength() > 10 || sink.bufferedAhead() > 60 || sink.totalBuffered() > 120)) {
          await new Promise((r) => setTimeout(r, 500));
        }
        if (destroyed || pumpToken !== myToken) return;
        sink.append(chunk.data);
        stats.bytesWritten += chunk.data.byteLength;
        stats.fragments++;
      }
    });
    const target = new mb.StreamTarget(writable);
    const output = new mb.Output({
      format: new mb.Mp4OutputFormat({ fastStart: "fragmented" }),
      target
    });
    const videoSource = new mb.EncodedVideoPacketSource(mbVideoCodec);
    output.addVideoTrack(videoSource);
    let audioSource = null;
    if (mbAudioCodec && inputAudio?.isAudioTrack()) {
      audioSource = new mb.EncodedAudioPacketSource(mbAudioCodec);
      output.addAudioTrack(audioSource);
    }
    currentOutput = output;
    return { output, videoSource, audioSource };
  }
  async function pumpLoop(token, fromTime) {
    const { output, videoSource, audioSource } = createOutput();
    await output.start();
    const startVideoPacket = fromTime > 0 ? await videoSink.getKeyPacket(fromTime) ?? await videoSink.getFirstPacket() : await videoSink.getFirstPacket();
    if (!startVideoPacket) return;
    const startAudioPacket = audioSink ? audioSink && fromTime > 0 ? await audioSink.getPacket(fromTime) ?? await audioSink.getFirstPacket() : await audioSink.getFirstPacket() : null;
    const videoIter = videoSink.packets(startVideoPacket);
    const audioIter = audioSink && startAudioPacket ? audioSink.packets(startAudioPacket) : null;
    let vNext = await videoIter.next();
    let aNext = audioIter ? await audioIter.next() : { done: true, value: void 0 };
    let firstVideo = true;
    let firstAudio = true;
    while (!destroyed && pumpToken === token && (!vNext.done || !aNext.done)) {
      while (!destroyed && pumpToken === token && sink && (sink.bufferedAhead() > 30 || sink.queueLength() > 20 || sink.totalBuffered() > 90)) {
        await new Promise((r) => setTimeout(r, 500));
      }
      if (destroyed || pumpToken !== token) break;
      const vTs = !vNext.done ? vNext.value.timestamp : Number.POSITIVE_INFINITY;
      const aTs = !aNext.done ? aNext.value.timestamp : Number.POSITIVE_INFINITY;
      const forceVideoFirst = firstVideo && !vNext.done;
      if (!vNext.done && (forceVideoFirst || vTs <= aTs)) {
        await videoSource.add(
          vNext.value,
          firstVideo && videoConfig ? { decoderConfig: videoConfig } : void 0
        );
        firstVideo = false;
        stats.videoPackets++;
        vNext = await videoIter.next();
      } else if (audioIter && audioSource && !aNext.done) {
        await audioSource.add(
          aNext.value,
          firstAudio && audioConfig ? { decoderConfig: audioConfig } : void 0
        );
        firstAudio = false;
        stats.audioPackets++;
        aNext = await audioIter.next();
      } else {
        break;
      }
    }
    if (!destroyed && pumpToken === token) {
      await output.finalize();
      sink?.endOfStream();
    }
  }
  return {
    async start(fromTime = 0, autoPlay = false) {
      pendingAutoPlay = autoPlay;
      pendingStartTime = fromTime;
      pumpLoop(++pumpToken, fromTime).catch((err) => {
        console.error("[avbridge] remux pipeline failed:", err);
        try {
          sink?.destroy();
        } catch {
        }
      });
    },
    async seek(time, autoPlay = false) {
      if (sink) {
        sink.setPlayOnSeek(autoPlay);
        sink.invalidate(time);
      } else {
        pendingAutoPlay = autoPlay;
        pendingStartTime = time;
      }
      pumpLoop(++pumpToken, time).catch((err) => {
        console.error("[avbridge] remux pipeline reseek failed:", err);
      });
    },
    setAutoPlay(autoPlay) {
      pendingAutoPlay = autoPlay;
      if (sink) sink.setPlayOnSeek(autoPlay);
    },
    async setAudioTrack(trackId, time, autoPlay) {
      if (selectedAudioTrackId === trackId) return;
      if (!ctx.audioTracks.some((t) => t.id === trackId)) {
        console.warn("[avbridge] remux: setAudioTrack \u2014 unknown track id", trackId);
        return;
      }
      pumpToken++;
      selectedAudioTrackId = trackId;
      await rebuildAudio().catch((err) => {
        console.warn("[avbridge] remux: rebuildAudio failed:", err.message);
      });
      if (sink) {
        try {
          sink.destroy();
        } catch {
        }
        sink = null;
      }
      pendingAutoPlay = autoPlay;
      pendingStartTime = time;
      pumpLoop(++pumpToken, time).catch((err) => {
        console.error("[avbridge] remux pipeline setAudioTrack pump failed:", err);
      });
    },
    async destroy() {
      destroyed = true;
      pumpToken++;
      try {
        if (currentOutput) await currentOutput.cancel();
      } catch {
      }
      try {
        await input.dispose();
      } catch {
      }
      sink?.destroy();
    },
    stats() {
      return { ...stats, decoderType: "remux" };
    }
  };
}

// src/strategies/remux/index.ts
async function createRemuxSession(context, video) {
  let pipeline;
  try {
    pipeline = await createRemuxPipeline(context, video);
  } catch (err) {
    throw new Error(
      `remux strategy failed to start: ${err.message}. The container or codec combination is not supported by mediabunny + MSE on this browser.`
    );
  }
  let started = false;
  let wantPlay = false;
  return {
    strategy: "remux",
    async play() {
      wantPlay = true;
      if (!started) {
        started = true;
        await pipeline.start(video.currentTime || 0, true);
        return;
      }
      pipeline.setAutoPlay(true);
      try {
        await video.play();
      } catch {
      }
    },
    pause() {
      wantPlay = false;
      video.pause();
    },
    async seek(time) {
      if (!started) {
        started = true;
        await pipeline.seek(time, wantPlay);
        return;
      }
      const wasPlaying = !video.paused;
      await pipeline.seek(time, wasPlaying || wantPlay);
      queueMicrotask(() => {
        try {
          video.dispatchEvent(new Event("seeked"));
        } catch {
        }
      });
    },
    async setAudioTrack(id) {
      if (!context.audioTracks.some((t) => t.id === id)) {
        console.warn("[avbridge] remux: setAudioTrack \u2014 unknown track id", id);
        return;
      }
      const wasPlaying = !video.paused;
      const time = video.currentTime || 0;
      if (!started) {
        started = true;
        await pipeline.setAudioTrack(id, time, wantPlay || wasPlaying);
        return;
      }
      await pipeline.setAudioTrack(id, time, wasPlaying || wantPlay);
    },
    async setSubtitleTrack(id) {
      const tracks = video.textTracks;
      for (let i = 0; i < tracks.length; i++) {
        tracks[i].mode = i === id ? "showing" : "disabled";
      }
    },
    getCurrentTime() {
      return video.currentTime || 0;
    },
    async destroy() {
      video.pause();
      await pipeline.destroy();
      video.removeAttribute("src");
      video.load();
    },
    getRuntimeStats() {
      return pipeline.stats();
    }
  };
}

// src/strategies/fallback/video-renderer.ts
function isDebug() {
  return typeof globalThis !== "undefined" && !!globalThis.AVBRIDGE_DEBUG;
}
var lastDebugLog = 0;
var VideoRenderer = class {
  constructor(target, clock, fps = 30) {
    this.target = target;
    this.clock = clock;
    this.paintIntervalMs = Math.max(1, 1e3 / fps);
    this.firstFrameReady = new Promise((resolve) => {
      this.resolveFirstFrame = resolve;
    });
    this.canvas = document.createElement("canvas");
    this.canvas.style.cssText = "position:absolute;left:0;top:0;width:100%;height:100%;background:black;object-fit:var(--avbridge-fit, contain);";
    const parent = target.parentElement ?? target.parentNode;
    if (parent && parent instanceof HTMLElement) {
      if (getComputedStyle(parent).position === "static") {
        parent.style.position = "relative";
      }
    }
    if (parent) {
      parent.insertBefore(this.canvas, target);
    } else {
      console.warn(
        "[avbridge] fallback renderer: target <video> has no parent; appending canvas to document.body as a fallback."
      );
      document.body.appendChild(this.canvas);
    }
    target.style.visibility = "hidden";
    const overlayParent = parent instanceof HTMLElement ? parent : document.body;
    this.subtitleOverlay = new chunkWRKO6Q42_cjs.SubtitleOverlay(overlayParent);
    this.watchTextTracks(target);
    const ctx = this.canvas.getContext("2d");
    if (!ctx) throw new Error("video renderer: failed to acquire 2D context");
    this.ctx = ctx;
    this.tick = this.tick.bind(this);
    this.rafHandle = requestAnimationFrame(this.tick);
  }
  target;
  clock;
  canvas;
  ctx;
  queue = [];
  rafHandle = null;
  destroyed = false;
  framesPainted = 0;
  framesDroppedLate = 0;
  framesDroppedOverflow = 0;
  /** True once the head frame has been painted as a pre-roll poster
   *  since the last flush. Used to ensure pre-roll paints exactly one
   *  frame (held static) during the post-seek discard window. */
  prerolled = false;
  /** PTS (µs) of the most recently painted frame. Used as the calibration
   *  reference on the first post-flush snap: the pre-roll path paints one
   *  frame *before* PTS-based playback starts, so the queue head's PTS at
   *  first PTS-based paint is the *next* frame, off by one frameDur from
   *  the actually-displayed frame. Calibrating against the painted frame
   *  instead of the queue head removes that one-frame offset and yields
   *  calib ≈ 0 instead of +frameDur. */
  lastPaintedPtsUs = 0;
  hasLastPaintedPts = false;
  /** Audio-clock reading (ms) at the previous paint, for overlay Δaud. */
  lastPaintAudMs = 0;
  /** Wall-clock time of the last paint, in ms (performance.now()). */
  lastPaintWall = 0;
  /** Minimum ms between paints — paces video at roughly source fps. */
  paintIntervalMs;
  /** Cumulative count of frames skipped because all PTS are in the future. */
  ticksWaiting = 0;
  /** Cumulative count of ticks where PTS mode painted a frame. */
  ticksPainted = 0;
  /**
   * Subtitle overlay div attached to the stage wrapper alongside the
   * canvas. Created lazily when subtitle tracks are attached via the
   * target's `<track>` children. Canvas strategies (hybrid, fallback)
   * hide the <video>, so we can't rely on the browser's native cue
   * rendering; we read TextTrack.cues and render into this overlay.
   */
  subtitleOverlay = null;
  subtitleTrack = null;
  /**
   * Calibration offset (microseconds) between video PTS and audio clock.
   * Video PTS and AudioContext.currentTime can drift ~0.1% relative to
   * each other (different clock domains). Over 45 minutes that's 2.6s.
   * We measure the offset on the first painted frame and update it
   * periodically so the PTS comparison stays calibrated.
   */
  ptsCalibrationUs = 0;
  ptsCalibrated = false;
  lastCalibrationWall = 0;
  /** Resolves once the first decoded frame has been enqueued. */
  firstFrameReady;
  resolveFirstFrame;
  /**
   * True once at least one frame has been enqueued *since the last flush*.
   * Used by `readyState` — initial cold-start reports HAVE_NOTHING until
   * any frame has arrived, and after a seek we want the same semantics
   * (HAVE_NOTHING until post-seek frames arrive), so the cumulative
   * `framesPainted > 0` that used to live here was wrong: it kept the
   * state "true forever" after the first frame ever, so post-seek
   * `waitForBuffer()` would exit immediately with an empty queue and
   * leave video frozen while audio kept going.
   */
  hasFrames() {
    return this.queue.length > 0 || this.hasEverEnqueuedSinceFlush;
  }
  hasEverEnqueuedSinceFlush = false;
  /** Current depth of the frame queue. Used by the decoder for backpressure. */
  queueDepth() {
    return this.queue.length;
  }
  /**
   * Cap the decoder may fill the queue up to. Used by the decoder's
   * enqueue-side discard logic (it closes new frames instead of pushing
   * them when this is reached). Sized so a long post-seek catch-up
   * fits — the decoder produces frames at PTS T_kf onwards rapidly
   * while the demuxer is chewing through pre-target audio; if the
   * queue can hold the whole post-seek burst, the renderer plays
   * smoothly from pre-roll without a frozen-video gap when audio.start
   * fires. At ~340 KB per SD frame the cap is ~85 MB peak; at HD it's
   * larger but still bounded.
   */
  queueHighWater = 256;
  enqueue(frame) {
    if (this.destroyed) {
      frame.close();
      return;
    }
    this.queue.push(frame);
    this.hasEverEnqueuedSinceFlush = true;
    if (this.queue.length === 1 && this.framesPainted === 0) {
      this.resolveFirstFrame();
    }
    while (this.queue.length > this.queueHighWater + 8) {
      this.queue.shift()?.close();
      this.framesDroppedOverflow++;
    }
  }
  /**
   * Watch the target <video>'s textTracks list. When a track is added,
   * grab it and start polling cues on each render tick. Existing tracks
   * (if any) are picked up immediately.
   */
  watchTextTracks(target) {
    const pick = () => {
      if (this.subtitleTrack) return;
      const tracks = target.textTracks;
      if (isDebug()) {
        console.log(`[avbridge:subs] watchTextTracks pick() \u2014 ${tracks.length} tracks`);
      }
      for (let i = 0; i < tracks.length; i++) {
        const t = tracks[i];
        if (isDebug()) {
          console.log(`[avbridge:subs] track ${i}: kind=${t.kind} mode=${t.mode} cues=${t.cues?.length ?? 0}`);
        }
        if (t.kind === "subtitles" || t.kind === "captions") {
          this.subtitleTrack = t;
          t.mode = "hidden";
          if (isDebug()) {
            console.log(`[avbridge:subs] picked track, mode=hidden`);
          }
          const trackEl = target.querySelector(`track[srclang="${t.language}"]`);
          if (trackEl) {
            trackEl.addEventListener("load", () => {
              if (isDebug()) {
                console.log(`[avbridge:subs] track element loaded, cues=${t.cues?.length ?? 0}`);
              }
            });
            trackEl.addEventListener("error", (ev) => {
              console.warn(`[avbridge:subs] track element error:`, ev);
            });
          }
          break;
        }
      }
    };
    pick();
    if (typeof target.textTracks.addEventListener === "function") {
      target.textTracks.addEventListener("addtrack", (e) => {
        if (isDebug()) {
          console.log("[avbridge:subs] addtrack event fired");
        }
        pick();
      });
    }
  }
  _loggedCues = false;
  /** Find the active cue (if any) for the given media time. */
  updateSubtitles() {
    if (!this.subtitleOverlay || !this.subtitleTrack) return;
    const cues = this.subtitleTrack.cues;
    if (!cues || cues.length === 0) return;
    if (isDebug() && !this._loggedCues) {
      this._loggedCues = true;
      console.log(`[avbridge:subs] cues available: ${cues.length}, first start=${cues[0].startTime}, last end=${cues[cues.length - 1].endTime}`);
    }
    const t = this.clock.now();
    let activeText = "";
    for (let i = 0; i < cues.length; i++) {
      const c = cues[i];
      if (t >= c.startTime && t <= c.endTime) {
        const vttCue = c;
        activeText = vttCue.text ?? "";
        break;
      }
    }
    this.subtitleOverlay.setText(activeText.replace(/<[^>]+>/g, ""));
  }
  tick() {
    if (this.destroyed) return;
    this.rafHandle = requestAnimationFrame(this.tick);
    this.updateSubtitles();
    if (this.queue.length === 0) return;
    const playing = this.clock.isPlaying();
    if (!playing) {
      if (!this.prerolled && this.queue.length > 0) {
        this.prerolled = true;
        this.paint(this.queue[0]);
      }
      return;
    }
    const rawAudioNowUs = this.clock.now() * 1e6;
    const headTs = this.queue[0].timestamp ?? 0;
    const hasPts = headTs > 0 || this.queue.length > 1;
    if (hasPts) {
      const wallNow2 = performance.now();
      if (!this.ptsCalibrated) {
        const anchorUs = (this.clock.anchorTime?.() ?? this.clock.now()) * 1e6;
        const referencePtsUs = this.hasLastPaintedPts ? this.lastPaintedPtsUs : headTs;
        this.ptsCalibrationUs = referencePtsUs - anchorUs;
        this.ptsCalibrated = true;
        this.lastCalibrationWall = wallNow2;
        if (isDebug()) {
          console.log(
            `[avbridge:renderer] CALIB-FIRST audioAnchor=${(anchorUs / 1e3).toFixed(1)}ms prerolledPTS=${this.hasLastPaintedPts ? (this.lastPaintedPtsUs / 1e3).toFixed(1) : "n/a"}ms queueHeadPTS=${(headTs / 1e3).toFixed(1)}ms rawAudioNow=${(rawAudioNowUs / 1e3).toFixed(1)}ms \u2192 calib=${(this.ptsCalibrationUs / 1e3).toFixed(1)}ms`
          );
        }
      } else if (wallNow2 - this.lastCalibrationWall > 1e4) {
        const oldCalib = this.ptsCalibrationUs;
        this.ptsCalibrationUs = headTs - rawAudioNowUs;
        this.lastCalibrationWall = wallNow2;
        if (isDebug()) {
          console.log(
            `[avbridge:renderer] CALIB-RESNAP headPTS=${(headTs / 1e3).toFixed(1)}ms rawAudioNow=${(rawAudioNowUs / 1e3).toFixed(1)}ms calib ${(oldCalib / 1e3).toFixed(1)}ms \u2192 ${(this.ptsCalibrationUs / 1e3).toFixed(1)}ms (\u0394=${((this.ptsCalibrationUs - oldCalib) / 1e3).toFixed(1)}ms after 10s)`
          );
        }
      }
      const audioNowUs = rawAudioNowUs + this.ptsCalibrationUs;
      const deadlineUs = audioNowUs;
      let bestIdx = -1;
      for (let i = 0; i < this.queue.length; i++) {
        const ts = this.queue[i].timestamp ?? 0;
        if (ts <= deadlineUs) {
          bestIdx = i;
        } else {
          break;
        }
      }
      if (bestIdx < 0) {
        this.ticksWaiting++;
        if (isDebug()) {
          const now = performance.now();
          if (now - lastDebugLog > 1e3) {
            const headPtsMs = (headTs / 1e3).toFixed(1);
            const audioMs = (audioNowUs / 1e3).toFixed(1);
            const rawDriftMs = ((headTs - rawAudioNowUs) / 1e3).toFixed(1);
            const calibMs = (this.ptsCalibrationUs / 1e3).toFixed(1);
            console.log(
              `[avbridge:renderer] WAIT q=${this.queue.length} headPTS=${headPtsMs}ms calibAudio=${audioMs}ms rawDrift=${rawDriftMs}ms calib=${calibMs}ms painted=${this.framesPainted} dropped=${this.framesDroppedLate}`
            );
            lastDebugLog = now;
          }
        }
        return;
      }
      const _relaxDrop = globalThis.AVBRIDGE_RELAX_DROP === true;
      let dropped = 0;
      const initialBestIdx = bestIdx;
      if (!_relaxDrop) {
        while (bestIdx > 0) {
          this.queue.shift()?.close();
          this.framesDroppedLate++;
          bestIdx--;
          dropped++;
        }
      }
      const paintTs = this.queue[0]?.timestamp ?? 0;
      if (isDebug()) {
        console.log(`[TRACE] PAINT bestIdx_initial=${initialBestIdx} dropped=${dropped} paintPts=${(paintTs / 1e3).toFixed(1)}ms audioNow=${(audioNowUs / 1e3).toFixed(1)}ms deadline=${(deadlineUs / 1e3).toFixed(1)}ms queueLen=${this.queue.length} wall=${performance.now().toFixed(0)}`);
      }
      this.ticksPainted++;
      if (isDebug()) {
        const now = performance.now();
        if (now - lastDebugLog > 1e3) {
          const paintedTs = this.queue[0]?.timestamp ?? 0;
          const audioMs = (audioNowUs / 1e3).toFixed(1);
          const ptsMs = (paintedTs / 1e3).toFixed(1);
          const rawDriftMs = ((paintedTs - rawAudioNowUs) / 1e3).toFixed(1);
          const calibMs = (this.ptsCalibrationUs / 1e3).toFixed(1);
          console.log(
            `[avbridge:renderer] PAINT q=${this.queue.length} calibAudio=${audioMs}ms nextPTS=${ptsMs}ms rawDrift=${rawDriftMs}ms calib=${calibMs}ms dropped=${dropped} total_drops=${this.framesDroppedLate} painted=${this.framesPainted}`
          );
          lastDebugLog = now;
        }
      }
      const frame2 = this.queue.shift();
      this.paint(frame2);
      frame2.close();
      this.lastPaintWall = performance.now();
      return;
    }
    const wallNow = performance.now();
    if (wallNow - this.lastPaintWall < this.paintIntervalMs - 2) return;
    const frame = this.queue.shift();
    this.paint(frame);
    frame.close();
    this.lastPaintWall = wallNow;
  }
  paint(frame) {
    if (this.canvas.width !== frame.displayWidth || this.canvas.height !== frame.displayHeight) {
      this.canvas.width = frame.displayWidth;
      this.canvas.height = frame.displayHeight;
    }
    try {
      this.ctx.drawImage(frame, 0, 0, this.canvas.width, this.canvas.height);
      if (isDebug()) {
        const wallNow = performance.now();
        const audNowMs = this.clock.now() * 1e3;
        const ptsMs = (frame.timestamp ?? 0) / 1e3;
        const dWall = this.lastPaintWall > 0 ? wallNow - this.lastPaintWall : 0;
        const dAud = this.lastPaintAudMs > 0 ? audNowMs - this.lastPaintAudMs : 0;
        const dPts = this.hasLastPaintedPts ? ptsMs - this.lastPaintedPtsUs / 1e3 : 0;
        this.ctx.save();
        this.ctx.font = "bold 18px monospace";
        const lines = [
          `#${this.framesPainted + 1}  pts=${ptsMs.toFixed(0)}  aud=${audNowMs.toFixed(0)}  wall=${wallNow.toFixed(0)}`,
          `\u0394pts=${dPts.toFixed(0)}  \u0394aud=${dAud.toFixed(0)}  \u0394wall=${dWall.toFixed(0)}`
        ];
        const lineHeight = 22;
        const padTop = 6;
        const stripH = padTop + lineHeight * lines.length;
        this.ctx.fillStyle = "rgba(0,0,0,0.7)";
        this.ctx.fillRect(0, 0, this.canvas.width, stripH);
        this.ctx.fillStyle = "#0f0";
        for (let i = 0; i < lines.length; i++) {
          this.ctx.fillText(lines[i], 8, padTop + lineHeight * (i + 1) - 4);
        }
        this.ctx.restore();
      }
      this.lastPaintedPtsUs = frame.timestamp ?? 0;
      this.hasLastPaintedPts = true;
      this.lastPaintAudMs = this.clock.now() * 1e3;
      this.framesPainted++;
    } catch (err) {
      if (this.framesPainted === 0 && this.framesDroppedLate === 0) {
        console.warn("[avbridge] canvas drawImage failed:", err);
      }
    }
  }
  /** Discard all queued frames. Used by seek to drop stale buffers. */
  flush() {
    const count = this.queue.length;
    while (this.queue.length > 0) this.queue.shift()?.close();
    this.prerolled = false;
    this.hasLastPaintedPts = false;
    this.ptsCalibrated = false;
    this.hasEverEnqueuedSinceFlush = false;
    if (isDebug() && count > 0) {
      console.log(`[avbridge:renderer] FLUSH discarded=${count} painted=${this.framesPainted} drops=${this.framesDroppedLate}`);
    }
  }
  stats() {
    let queueSpanMs = 0;
    let queueHeadMs = 0;
    let queueTailMs = 0;
    if (this.queue.length > 0) {
      queueHeadMs = Math.round((this.queue[0].timestamp ?? 0) / 1e3);
      queueTailMs = Math.round((this.queue[this.queue.length - 1].timestamp ?? 0) / 1e3);
      queueSpanMs = Math.max(0, queueTailMs - queueHeadMs);
    }
    return {
      framesPainted: this.framesPainted,
      framesDroppedLate: this.framesDroppedLate,
      framesDroppedOverflow: this.framesDroppedOverflow,
      queueDepth: this.queue.length,
      queueHeadMs,
      queueTailMs,
      queueSpanMs
    };
  }
  destroy() {
    this.destroyed = true;
    if (this.rafHandle != null) cancelAnimationFrame(this.rafHandle);
    this.flush();
    if (this.subtitleOverlay) {
      this.subtitleOverlay.destroy();
      this.subtitleOverlay = null;
    }
    this.subtitleTrack = null;
    this.canvas.remove();
    this.target.style.visibility = "";
  }
};

// src/strategies/fallback/audio-output.ts
function isDebug2() {
  return typeof globalThis !== "undefined" && !!globalThis.AVBRIDGE_DEBUG;
}
var AudioOutput = class {
  ctx;
  gain;
  state = "idle";
  /**
   * Wall-clock fallback mode. When true, this output behaves as if audio
   * is unavailable — `now()` advances from `performance.now()` instead of
   * the audio context, `schedule()` is a no-op, and `bufferAhead()` returns
   * Infinity so the session's `waitForBuffer()` doesn't block on audio.
   *
   * Set by the decoder via {@link setNoAudio} when audio decode init fails.
   * This is what lets video play even when the audio codec isn't supported
   * by the loaded libav variant.
   */
  noAudio = false;
  /** Wall-clock anchor (ms from `performance.now()`) for noAudio mode. */
  wallAnchorMs = 0;
  /** Media time at which the next sample will be scheduled. */
  mediaTimeOfNext = 0;
  /** Anchor: media time `mediaTimeOfAnchor` corresponds to ctx time `ctxTimeAtAnchor`. */
  mediaTimeOfAnchor = 0;
  /**
   * Ctx time at which the first audible chunk will start playing. `-1`
   * before any chunk has been scheduled successfully (clock is frozen);
   * the actual ctx time once one has. The renderer's `clock.now()` uses
   * this to avoid advancing during the silent-gap window between
   * `audio.start()` and the first chunk that schedules without being
   * dropped — that gap is what produces the "audio-less fast-forward"
   * the user sees post-seek when the gate releases on video-only grace.
   */
  firstAudibleCtxStart = -1;
  ctxTimeAtAnchor = 0;
  pendingQueue = [];
  framesScheduled = 0;
  destroyed = false;
  /** User-set volume (0..1). Applied to the gain node. */
  _volume = 1;
  /** User-set muted flag. When true, gain is forced to 0. */
  _muted = false;
  /** Playback rate. Scales the media clock and each AudioBufferSourceNode's
   *  playbackRate so audio pitches up/down accordingly (same as native
   *  <video>.playbackRate). Default 1. */
  _rate = 1;
  constructor() {
    this.ctx = new AudioContext();
    this.gain = this.ctx.createGain();
    this.gain.connect(this.ctx.destination);
  }
  /** Set volume (0..1). Applied immediately to the gain node. */
  setVolume(v) {
    this._volume = Math.max(0, Math.min(1, v));
    this.applyGain();
  }
  getVolume() {
    return this._volume;
  }
  /** Set muted. When true, output is silenced regardless of volume. */
  setMuted(m) {
    this._muted = m;
    this.applyGain();
  }
  getMuted() {
    return this._muted;
  }
  /** Set playback rate. Scales the media clock and pitches audio output
   *  (same as native <video>.playbackRate — speed without pitch correction).
   *  Rebases the anchor so the clock transition is seamless. */
  setPlaybackRate(rate) {
    if (rate === this._rate) return;
    const t = this.now();
    this.mediaTimeOfAnchor = t;
    this.ctxTimeAtAnchor = this.ctx.currentTime;
    this.wallAnchorMs = performance.now();
    this._rate = rate;
  }
  getPlaybackRate() {
    return this._rate;
  }
  applyGain() {
    const target = this._muted ? 0 : this._volume;
    try {
      this.gain.gain.value = target;
    } catch {
    }
  }
  /**
   * Switch into wall-clock fallback mode. Called by the decoder when no
   * audio decoder could be initialized for the source. Once set, this
   * output drives playback time from `performance.now()` and ignores
   * any incoming audio samples.
   */
  setNoAudio() {
    this.noAudio = true;
  }
  // ── ClockSource ────────────────────────────────────────────────────────
  now() {
    if (this.noAudio) {
      if (this.state === "playing") {
        return this.mediaTimeOfAnchor + (performance.now() - this.wallAnchorMs) / 1e3 * this._rate;
      }
      return this.mediaTimeOfAnchor;
    }
    if (this.state === "playing") {
      if (this.firstAudibleCtxStart < 0) {
        return this.mediaTimeOfAnchor;
      }
      return this.mediaTimeOfAnchor + (this.ctx.currentTime - this.ctxTimeAtAnchor) * this._rate;
    }
    return this.mediaTimeOfAnchor;
  }
  anchorTime() {
    return this.mediaTimeOfAnchor;
  }
  isPlaying() {
    return this.state === "playing";
  }
  // ── Buffering ─────────────────────────────────────────────────────────
  /**
   * How many seconds of audio are buffered ahead of the current playback
   * position. While idle, this counts the pending queue. While playing,
   * it counts how far `mediaTimeOfNext` is ahead of `now()`.
   */
  bufferAhead() {
    if (this.noAudio) return 0;
    if (this.state === "idle") {
      let sec = 0;
      for (const c of this.pendingQueue) sec += c.durationSec;
      return sec;
    }
    return Math.max(0, this.mediaTimeOfNext - this.now());
  }
  /** True if this output is in wall-clock fallback mode (no audio decode). */
  isNoAudio() {
    return this.noAudio;
  }
  /**
   * Schedule a chunk of decoded samples. Queues internally while idle (cold
   * start or post-seek), schedules directly to the audio graph while playing.
   * In wall-clock mode, samples are silently discarded.
   *
   * `ptsSec` is the chunk's source-domain content PTS in seconds, from
   * the demuxer. When provided, the chunk plays at the ctx-time
   * corresponding to that PTS — so pre-target audio after a seek
   * naturally drops (its computed `ctxStart` falls in the past) and
   * post-target audio plays at its true content time, without any
   * external trim or anchor rebase. When `ptsSec` is null (cold start
   * with no PTS yet, or codecs whose packet→frame mapping isn't 1:1),
   * the chunk is scheduled sequentially after `mediaTimeOfNext` — the
   * pre-refactor behavior.
   */
  schedule(samples, channels, sampleRate, ptsSec) {
    if (this.destroyed || this.noAudio) return;
    const frameCount = samples.length / channels;
    const durationSec = frameCount / sampleRate;
    const hasPts = ptsSec != null && Number.isFinite(ptsSec);
    if (hasPts && ptsSec + durationSec / this._rate < this.mediaTimeOfAnchor) {
      return;
    }
    if (this.state === "idle" || this.state === "paused") {
      this.pendingQueue.push({
        samples,
        channels,
        sampleRate,
        frameCount,
        durationSec,
        ptsSec: hasPts ? ptsSec : null
      });
      return;
    }
    this.scheduleNow(
      samples,
      channels,
      sampleRate,
      frameCount,
      hasPts ? ptsSec : null
    );
  }
  scheduleNow(samples, channels, sampleRate, frameCount, ptsSec) {
    const durationSec = frameCount / sampleRate;
    let ctxStart;
    if (ptsSec != null) {
      ctxStart = this.ctxTimeAtAnchor + (ptsSec - this.mediaTimeOfAnchor) / this._rate;
      if (isDebug2()) {
        console.log(`[TRACE-AUD] PTS sched #${this.framesScheduled} pts=${ptsSec.toFixed(3)} dur=${durationSec.toFixed(4)} ctxStart=${ctxStart.toFixed(4)} ctxNow=${this.ctx.currentTime.toFixed(4)} anchor=${this.mediaTimeOfAnchor.toFixed(3)} ctxAnchor=${this.ctxTimeAtAnchor.toFixed(4)} mtNext=${this.mediaTimeOfNext.toFixed(3)} rate=${this._rate}`);
      }
      if (ctxStart < this.ctx.currentTime - 1e-3) {
        if (isDebug2()) {
          console.log(`[TRACE-AUD] DROP late chunk pts=${ptsSec.toFixed(3)} ctxStart=${ctxStart.toFixed(4)} < ctxNow=${this.ctx.currentTime.toFixed(4)}`);
        }
        return;
      }
      if (this.firstAudibleCtxStart < 0) {
        this.firstAudibleCtxStart = ctxStart;
        this.mediaTimeOfAnchor = ptsSec;
        this.ctxTimeAtAnchor = ctxStart;
        if (isDebug2()) {
          console.log(`[TRACE-AUD] UNFREEZE clock \u2014 first audible chunk pts=${ptsSec.toFixed(3)} ctxStart=${ctxStart.toFixed(4)} \u2192 anchor=${this.mediaTimeOfAnchor.toFixed(3)} ctxAnchor=${this.ctxTimeAtAnchor.toFixed(4)}`);
        }
      }
      const endMediaTime = ptsSec + durationSec / this._rate;
      if (endMediaTime > this.mediaTimeOfNext) {
        this.mediaTimeOfNext = endMediaTime;
      }
    } else {
      ctxStart = this.ctxTimeAtAnchor + (this.mediaTimeOfNext - this.mediaTimeOfAnchor) / this._rate;
      console.warn(`[TRACE-AUD] LEGACY (no PTS) sched dur=${durationSec.toFixed(4)} ctxStart=${ctxStart.toFixed(4)} ctxNow=${this.ctx.currentTime.toFixed(4)}`);
      if (ctxStart < this.ctx.currentTime) {
        console.warn(`[TRACE-AUD] REBASE anchor was=${this.mediaTimeOfAnchor.toFixed(3)} ctxAnchor was=${this.ctxTimeAtAnchor.toFixed(4)} \u2192 anchor=${this.mediaTimeOfNext.toFixed(3)} ctxAnchor=${this.ctx.currentTime.toFixed(4)}`);
        this.ctxTimeAtAnchor = this.ctx.currentTime;
        this.mediaTimeOfAnchor = this.mediaTimeOfNext;
        ctxStart = this.ctx.currentTime;
      }
      this.mediaTimeOfNext += durationSec;
    }
    const buffer = this.ctx.createBuffer(channels, frameCount, sampleRate);
    for (let ch = 0; ch < channels; ch++) {
      const channelData = buffer.getChannelData(ch);
      for (let i = 0; i < frameCount; i++) {
        channelData[i] = samples[i * channels + ch];
      }
    }
    const node = this.ctx.createBufferSource();
    node.buffer = buffer;
    node.connect(this.gain);
    if (this._rate !== 1) node.playbackRate.value = this._rate;
    node.start(ctxStart);
    this.framesScheduled++;
  }
  // ── Lifecycle ─────────────────────────────────────────────────────────
  /**
   * Start (or resume) playback. On a cold start (or after a reset), drains
   * the pending queue scheduling all queued samples to play starting at
   * `ctx.currentTime + STARTUP_DELAY`. On resume from pause, just re-anchors
   * the media↔ctx time mapping and unsuspends the context.
   */
  async start() {
    if (this.destroyed || this.state === "playing") return;
    if (this.noAudio) {
      this.wallAnchorMs = performance.now();
      this.state = "playing";
      return;
    }
    if (this.ctx.state === "suspended") {
      await this.ctx.resume();
    }
    try {
      this.gain.connect(this.ctx.destination);
    } catch {
    }
    if (this.state === "paused") {
      if (isDebug2()) {
        console.log(`[TRACE-AUD] START(resume) anchor=${this.mediaTimeOfAnchor.toFixed(3)} ctxAnchor=${this.ctxTimeAtAnchor.toFixed(4)} \u2192 ctxAnchor=${this.ctx.currentTime.toFixed(4)} ctxNow=${this.ctx.currentTime.toFixed(4)} pendingCount=${this.pendingQueue.length}`);
      }
      this.ctxTimeAtAnchor = this.ctx.currentTime;
      this.state = "playing";
      const drain2 = this.pendingQueue;
      this.pendingQueue = [];
      for (const c of drain2) {
        this.scheduleNow(c.samples, c.channels, c.sampleRate, c.frameCount, c.ptsSec);
      }
      return;
    }
    const STARTUP_DELAY = 0.05;
    this.ctxTimeAtAnchor = this.ctx.currentTime + STARTUP_DELAY;
    this.mediaTimeOfNext = this.mediaTimeOfAnchor;
    this.state = "playing";
    if (isDebug2()) {
      console.log(`[TRACE-AUD] START(cold) anchor=${this.mediaTimeOfAnchor.toFixed(3)} ctxAnchor=${this.ctxTimeAtAnchor.toFixed(4)} mtNext=${this.mediaTimeOfNext.toFixed(3)} ctxNow=${this.ctx.currentTime.toFixed(4)} pendingCount=${this.pendingQueue.length}`);
    }
    const drain = this.pendingQueue;
    this.pendingQueue = [];
    for (const c of drain) {
      this.scheduleNow(c.samples, c.channels, c.sampleRate, c.frameCount, c.ptsSec);
    }
  }
  /** Pause playback. Suspends the audio context. */
  async pause() {
    if (this.state !== "playing") return;
    this.mediaTimeOfAnchor = this.now();
    this.state = "paused";
    if (this.noAudio) return;
    try {
      this.gain.disconnect();
    } catch {
    }
    if (this.ctx.state === "running") {
      await this.ctx.suspend();
    }
  }
  /**
   * Reset to a new media time. Discards all queued and scheduled audio,
   * disconnects the gain node so any in-flight scheduled buffers are cut
   * off, and returns to the idle state. Used by `seek()`.
   *
   * After reset, callers should re-buffer audio (the decoder will start
   * supplying new samples) and then call `start()` to resume playback.
   */
  async reset(newMediaTime) {
    if (isDebug2()) {
      console.log(`[TRACE-AUD] RESET to=${newMediaTime.toFixed(3)} prev_anchor=${this.mediaTimeOfAnchor.toFixed(3)} prev_mtNext=${this.mediaTimeOfNext.toFixed(3)} prev_ctxAnchor=${this.ctxTimeAtAnchor.toFixed(4)} ctxNow=${this.ctx.currentTime.toFixed(4)} state=${this.state}`);
    }
    if (this.noAudio) {
      this.pendingQueue = [];
      this.mediaTimeOfAnchor = newMediaTime;
      this.wallAnchorMs = performance.now();
      this.state = "idle";
      return;
    }
    try {
      this.gain.disconnect();
    } catch {
    }
    this.gain = this.ctx.createGain();
    this.gain.connect(this.ctx.destination);
    this.applyGain();
    this.pendingQueue = [];
    this.mediaTimeOfAnchor = newMediaTime;
    this.mediaTimeOfNext = newMediaTime;
    this.ctxTimeAtAnchor = this.ctx.currentTime;
    this.firstAudibleCtxStart = -1;
    this.state = "idle";
    if (this.ctx.state === "running") {
      await this.ctx.suspend();
    }
  }
  stats() {
    return {
      framesScheduled: this.framesScheduled,
      bufferAhead: this.bufferAhead(),
      audioState: this.state,
      clockMode: this.noAudio ? "wall" : "audio"
    };
  }
  destroy() {
    this.destroyed = true;
    try {
      this.ctx.close();
    } catch {
    }
  }
};

// src/strategies/hybrid/decoder.ts
async function startHybridDecoder(opts) {
  const variant = chunkF3LQJKXK_cjs.pickLibavVariant(opts.context);
  const libav = await chunkG4APZMCP_cjs.loadLibav(variant);
  const bridge = await loadBridge();
  const { prepareLibavInput } = await import('./libav-http-reader-Q356EO2K.cjs');
  const inputHandle = await prepareLibavInput(libav, opts.filename, opts.source, opts.transport);
  const readPkt = await libav.av_packet_alloc();
  const [fmt_ctx, streams] = await libav.ff_init_demuxer_file(opts.filename);
  const videoStream = streams.find((s) => s.codec_type === libav.AVMEDIA_TYPE_VIDEO) ?? null;
  const firstAudioTrackId = opts.context.audioTracks[0]?.id;
  let audioStream = (firstAudioTrackId != null ? streams.find((s) => s.codec_type === libav.AVMEDIA_TYPE_AUDIO && s.index === firstAudioTrackId) : void 0) ?? streams.find((s) => s.codec_type === libav.AVMEDIA_TYPE_AUDIO) ?? null;
  if (!videoStream && !audioStream) {
    throw new Error("hybrid decoder: file has no decodable streams");
  }
  let fatalHandler = null;
  let fatalFired = false;
  function fireFatal(reason) {
    if (fatalFired) return;
    fatalFired = true;
    fatalHandler?.(reason);
  }
  let videoDecoder = null;
  let videoTimeBase;
  if (videoStream) {
    try {
      const config = await bridge.videoStreamToConfig(libav, videoStream);
      if (!config) throw new Error("bridge returned null config");
      const supported = await VideoDecoder.isConfigSupported(config);
      if (!supported.supported) throw new Error(`VideoDecoder does not support config: ${JSON.stringify(config)}`);
      videoDecoder = new VideoDecoder({
        output: (frame) => {
          opts.renderer.enqueue(frame);
          videoFramesDecoded++;
        },
        error: (err) => {
          console.error("[avbridge] WebCodecs VideoDecoder error:", err);
          fireFatal(`WebCodecs VideoDecoder error: ${err.message}`);
        }
      });
      videoDecoder.configure(config);
      if (videoStream.time_base_num && videoStream.time_base_den) {
        videoTimeBase = [videoStream.time_base_num, videoStream.time_base_den];
      }
    } catch (err) {
      console.error("[avbridge] hybrid: failed to init WebCodecs VideoDecoder:", err);
      fireFatal(`WebCodecs VideoDecoder init failed: ${err.message}`);
      await inputHandle.detach().catch(() => {
      });
      throw err;
    }
  }
  let audioDec = null;
  let audioTimeBase;
  if (audioStream) {
    try {
      const [, c, pkt, frame] = await libav.ff_init_decoder(audioStream.codec_id, {
        codecpar: audioStream.codecpar
      });
      audioDec = { c, pkt, frame };
      if (audioStream.time_base_num && audioStream.time_base_den) {
        audioTimeBase = [audioStream.time_base_num, audioStream.time_base_den];
      }
    } catch (err) {
      console.warn(
        "[avbridge] hybrid: audio decoder unavailable for this codec \u2014 playing video with wall-clock timing:",
        err.message
      );
    }
  }
  if (!audioDec) {
    opts.audio.setNoAudio();
  }
  if (!videoDecoder && !audioDec) {
    await inputHandle.detach().catch(() => {
    });
    throw new Error("hybrid decoder: could not initialize any decoders");
  }
  let bsfCtx = null;
  let bsfPkt = null;
  let bsfRequiredButMissing = false;
  if (videoStream && opts.context.videoTracks[0]?.codec === "mpeg4") {
    try {
      bsfCtx = await libav.av_bsf_list_parse_str_js("mpeg4_unpack_bframes");
      if (bsfCtx != null && bsfCtx >= 0) {
        const parIn = await libav.AVBSFContext_par_in(bsfCtx);
        await libav.avcodec_parameters_copy(parIn, videoStream.codecpar);
        await libav.av_bsf_init(bsfCtx);
        bsfPkt = await libav.av_packet_alloc();
        chunkG4APZMCP_cjs.dbg.info("bsf", "mpeg4_unpack_bframes BSF active (hybrid)");
      } else {
        bsfRequiredButMissing = true;
        bsfCtx = null;
      }
    } catch (err) {
      bsfRequiredButMissing = true;
      bsfCtx = null;
      bsfPkt = null;
      chunkG4APZMCP_cjs.dbg.warn("bsf", `hybrid: mpeg4_unpack_bframes BSF init failed: ${err.message}`);
    }
    if (bsfRequiredButMissing) {
      console.error(
        "[avbridge] MPEG-4 Part 2 (DivX/Xvid) detected but mpeg4_unpack_bframes BSF is unavailable in this libav variant. Files with packed B-frames will play with incorrect frame ordering. Rebuild the libav variant with the `avbsf` fragment included."
      );
    }
  }
  async function applyBSF(packets) {
    if (!bsfCtx || !bsfPkt) return packets;
    const out = [];
    for (const pkt of packets) {
      await libav.ff_copyin_packet(bsfPkt, pkt);
      const sendErr = await libav.av_bsf_send_packet(bsfCtx, bsfPkt);
      if (sendErr < 0) {
        continue;
      }
      while (true) {
        const recvErr = await libav.av_bsf_receive_packet(bsfCtx, bsfPkt);
        if (recvErr < 0) break;
        out.push(await libav.ff_copyout_packet(bsfPkt));
      }
    }
    return out;
  }
  async function flushBSF() {
    if (!bsfCtx || !bsfPkt) return;
    try {
      if (libav.av_bsf_flush) {
        await libav.av_bsf_flush(bsfCtx);
      } else {
        while (true) {
          const err = await libav.av_bsf_receive_packet(bsfCtx, bsfPkt);
          if (err < 0) break;
        }
      }
    } catch {
    }
  }
  let destroyed = false;
  let pumpToken = 0;
  let pumpRunning = null;
  let packetsRead = 0;
  let videoFramesDecoded = 0;
  let audioFramesDecoded = 0;
  let videoChunksFed = 0;
  let bufferedUntilSec = 0;
  let syntheticVideoUs = 0;
  const videoTrackInfo = opts.context.videoTracks.find((t) => t.id === videoStream?.index);
  const videoFps = videoTrackInfo?.fps && videoTrackInfo.fps > 0 ? videoTrackInfo.fps : 30;
  const videoFrameStepUs = Math.max(1, Math.round(1e6 / videoFps));
  async function pumpLoop(myToken) {
    while (!destroyed && myToken === pumpToken) {
      let readErr;
      let packets;
      try {
        [readErr, packets] = await libav.ff_read_frame_multi(fmt_ctx, readPkt, {
          limit: 16 * 1024
        });
      } catch (err) {
        console.error("[avbridge] hybrid ff_read_frame_multi failed:", err);
        return;
      }
      if (myToken !== pumpToken || destroyed) return;
      const videoPackets = videoStream ? packets[videoStream.index] : void 0;
      const audioPackets = audioStream ? packets[audioStream.index] : void 0;
      if (videoPackets && videoTimeBase) {
        for (const pkt of videoPackets) {
          const sec = chunkHZUVMXBN_cjs.packetPtsSec(pkt, videoTimeBase);
          if (sec != null && sec > bufferedUntilSec) bufferedUntilSec = sec;
        }
      }
      if (audioPackets && audioTimeBase) {
        for (const pkt of audioPackets) {
          const sec = chunkHZUVMXBN_cjs.packetPtsSec(pkt, audioTimeBase);
          if (sec != null && sec > bufferedUntilSec) bufferedUntilSec = sec;
        }
      }
      if (audioDec && audioPackets && audioPackets.length > 0) {
        await decodeAudioBatch(
          audioPackets,
          myToken,
          /*flush*/
          false,
          audioTimeBase
        );
      }
      if (myToken !== pumpToken || destroyed) return;
      await new Promise((r) => setTimeout(r, 0));
      if (myToken !== pumpToken || destroyed) return;
      if (videoDecoder && videoPackets && videoPackets.length > 0) {
        const processed = await applyBSF(videoPackets);
        for (const pkt of processed) {
          if (myToken !== pumpToken || destroyed) return;
          chunkHZUVMXBN_cjs.sanitizePacketTimestamp(pkt, () => {
            const ts = syntheticVideoUs;
            syntheticVideoUs += videoFrameStepUs;
            return ts;
          }, videoTimeBase);
          try {
            const chunk = bridge.packetToEncodedVideoChunk(pkt, videoStream);
            videoDecoder.decode(chunk);
            videoChunksFed++;
          } catch (err) {
            if (videoChunksFed === 0) {
              console.warn("[avbridge] hybrid: packetToEncodedVideoChunk failed:", err);
              fireFatal(`WebCodecs chunk creation failed: ${err.message}`);
              return;
            }
          }
        }
      }
      packetsRead += (videoPackets?.length ?? 0) + (audioPackets?.length ?? 0);
      while (!destroyed && myToken === pumpToken && (videoDecoder && videoDecoder.decodeQueueSize > 10 || opts.audio.bufferAhead() > 2 || opts.renderer.queueDepth() >= opts.renderer.queueHighWater)) {
        await new Promise((r) => setTimeout(r, 50));
      }
      if (readErr === libav.AVERROR_EOF) {
        if (videoDecoder && videoDecoder.state === "configured") {
          try {
            await videoDecoder.flush();
          } catch {
          }
        }
        if (audioDec) await decodeAudioBatch([], myToken, true);
        return;
      }
      if (readErr && readErr !== 0 && readErr !== -libav.EAGAIN) {
        console.warn("[avbridge] hybrid ff_read_frame_multi returned", readErr);
        return;
      }
    }
  }
  async function decodeAudioBatch(pkts, myToken, flush = false, tb) {
    if (!audioDec || destroyed || myToken !== pumpToken) return;
    const pktPtsSec = pkts.map(
      (p) => tb ? chunkHZUVMXBN_cjs.packetPtsSec(p, tb) : null
    );
    const AUDIO_SUB_BATCH = 4;
    let allFrames = [];
    for (let i = 0; i < pkts.length; i += AUDIO_SUB_BATCH) {
      if (myToken !== pumpToken || destroyed) return;
      const slice = pkts.slice(i, i + AUDIO_SUB_BATCH);
      const isLast = i + AUDIO_SUB_BATCH >= pkts.length;
      try {
        const frames2 = await libav.ff_decode_multi(
          audioDec.c,
          audioDec.pkt,
          audioDec.frame,
          slice,
          isLast && flush ? { fin: true, ignoreErrors: true } : { ignoreErrors: true }
        );
        allFrames = allFrames.concat(frames2);
      } catch (err) {
        console.error("[avbridge] hybrid audio decode failed:", err);
        return;
      }
      if (!isLast) await new Promise((r) => setTimeout(r, 0));
    }
    if (pkts.length === 0 && flush) {
      try {
        allFrames = await libav.ff_decode_multi(
          audioDec.c,
          audioDec.pkt,
          audioDec.frame,
          [],
          { fin: true, ignoreErrors: true }
        );
      } catch (err) {
        console.error("[avbridge] hybrid audio flush failed:", err);
        return;
      }
    }
    if (myToken !== pumpToken || destroyed) return;
    const frames = allFrames;
    for (let i = 0; i < frames.length; i++) {
      if (myToken !== pumpToken || destroyed) return;
      const f = frames[i];
      const samples = chunkHZUVMXBN_cjs.libavFrameToInterleavedFloat32(f);
      if (samples) {
        const pts = pktPtsSec[i] ?? null;
        opts.audio.schedule(samples.data, samples.channels, samples.sampleRate, pts);
        audioFramesDecoded++;
      }
    }
  }
  pumpToken = 1;
  pumpRunning = pumpLoop(pumpToken).catch(
    (err) => console.error("[avbridge] hybrid pump failed:", err)
  );
  return {
    onFatalError(handler) {
      fatalHandler = handler;
      if (fatalFired) handler("WebCodecs decode failed (error occurred before handler attached)");
    },
    async destroy() {
      destroyed = true;
      pumpToken++;
      try {
        await pumpRunning;
      } catch {
      }
      try {
        if (bsfCtx) await libav.av_bsf_free(bsfCtx);
      } catch {
      }
      try {
        if (bsfPkt) await libav.av_packet_free?.(bsfPkt);
      } catch {
      }
      try {
        if (videoDecoder && videoDecoder.state !== "closed") videoDecoder.close();
      } catch {
      }
      try {
        if (audioDec) await libav.ff_free_decoder?.(audioDec.c, audioDec.pkt, audioDec.frame);
      } catch {
      }
      try {
        await libav.av_packet_free?.(readPkt);
      } catch {
      }
      try {
        await libav.avformat_close_input_js(fmt_ctx);
      } catch {
      }
      try {
        await inputHandle.detach();
      } catch {
      }
    },
    async setAudioTrack(trackId, timeSec) {
      if (audioStream && audioStream.index === trackId) return;
      const newStream = streams.find(
        (s) => s.codec_type === libav.AVMEDIA_TYPE_AUDIO && s.index === trackId
      );
      if (!newStream) {
        console.warn("[avbridge] hybrid: setAudioTrack \u2014 no stream with id", trackId);
        return;
      }
      const newToken = ++pumpToken;
      if (pumpRunning) {
        try {
          await pumpRunning;
        } catch {
        }
      }
      if (destroyed) return;
      if (audioDec) {
        try {
          await libav.ff_free_decoder?.(audioDec.c, audioDec.pkt, audioDec.frame);
        } catch {
        }
        audioDec = null;
      }
      try {
        const [, c, pkt, frame] = await libav.ff_init_decoder(newStream.codec_id, {
          codecpar: newStream.codecpar
        });
        audioDec = { c, pkt, frame };
        audioTimeBase = newStream.time_base_num && newStream.time_base_den ? [newStream.time_base_num, newStream.time_base_den] : void 0;
      } catch (err) {
        console.warn(
          "[avbridge] hybrid: setAudioTrack init failed \u2014 switching to no-audio:",
          err.message
        );
        audioDec = null;
        opts.audio.setNoAudio();
      }
      audioStream = newStream;
      try {
        const tsUs = Math.floor(timeSec * 1e6);
        const [tsLo, tsHi] = libav.f64toi64 ? libav.f64toi64(tsUs) : [tsUs | 0, Math.floor(tsUs / 4294967296)];
        await libav.av_seek_frame(
          fmt_ctx,
          -1,
          tsLo,
          tsHi,
          libav.AVSEEK_FLAG_BACKWARD ?? 0
        );
      } catch (err) {
        console.warn("[avbridge] hybrid: setAudioTrack seek failed:", err);
      }
      try {
        if (videoDecoder && videoDecoder.state === "configured") {
          await videoDecoder.flush();
        }
      } catch {
      }
      await flushBSF();
      syntheticVideoUs = Math.round(timeSec * 1e6);
      pumpRunning = pumpLoop(newToken).catch(
        (err) => console.error("[avbridge] hybrid pump failed (post-setAudioTrack):", err)
      );
    },
    async seek(timeSec) {
      const newToken = ++pumpToken;
      if (pumpRunning) {
        try {
          await pumpRunning;
        } catch {
        }
      }
      if (destroyed) return;
      try {
        const tsUs = Math.floor(timeSec * 1e6);
        const [tsLo, tsHi] = libav.f64toi64 ? libav.f64toi64(tsUs) : [tsUs | 0, Math.floor(tsUs / 4294967296)];
        await libav.av_seek_frame(
          fmt_ctx,
          -1,
          tsLo,
          tsHi,
          libav.AVSEEK_FLAG_BACKWARD ?? 0
        );
      } catch (err) {
        console.warn("[avbridge] hybrid av_seek_frame failed:", err);
      }
      try {
        if (videoDecoder && videoDecoder.state === "configured") {
          await videoDecoder.flush();
        }
      } catch {
      }
      try {
        if (audioDec) await libav.avcodec_flush_buffers?.(audioDec.c);
      } catch {
      }
      await flushBSF();
      syntheticVideoUs = Math.round(timeSec * 1e6);
      pumpRunning = pumpLoop(newToken).catch(
        (err) => console.error("[avbridge] hybrid pump failed (post-seek):", err)
      );
    },
    bufferedUntilSec() {
      return bufferedUntilSec;
    },
    stats() {
      return {
        decoderType: "webcodecs-hybrid",
        packetsRead,
        videoFramesDecoded,
        videoChunksFed,
        audioFramesDecoded,
        bsfApplied: bsfCtx ? ["mpeg4_unpack_bframes"] : [],
        bsfMissing: bsfRequiredButMissing ? ["mpeg4_unpack_bframes"] : [],
        videoDecodeQueueSize: videoDecoder?.decodeQueueSize ?? 0,
        // Confirmed transport info — see fallback decoder for the pattern.
        _transport: inputHandle.transport === "http-range" ? "http-range" : "memory",
        _rangeSupported: inputHandle.transport === "http-range",
        ...opts.renderer.stats(),
        ...opts.audio.stats()
      };
    }
  };
}
async function loadBridge() {
  try {
    const wrapper = await import('./libav-import-2ZVKV2E7.cjs');
    return wrapper.libavBridge;
  } catch (err) {
    throw new Error(
      `failed to load libavjs-webcodecs-bridge: ${err.message}`
    );
  }
}

// src/util/time-ranges.ts
function makeTimeRanges(ranges) {
  const frozen = ranges.slice();
  const impl = {
    get length() {
      return frozen.length;
    },
    start(index) {
      if (index < 0 || index >= frozen.length) {
        throw new DOMException(
          `TimeRanges.start: index ${index} out of range (length=${frozen.length})`,
          "IndexSizeError"
        );
      }
      return frozen[index][0];
    },
    end(index) {
      if (index < 0 || index >= frozen.length) {
        throw new DOMException(
          `TimeRanges.end: index ${index} out of range (length=${frozen.length})`,
          "IndexSizeError"
        );
      }
      return frozen[index][1];
    }
  };
  return impl;
}

// src/strategies/hybrid/index.ts
var READY_AUDIO_BUFFER_SECONDS = 0.3;
var READY_TIMEOUT_SECONDS = 10;
async function createHybridSession(ctx, target, transport) {
  const { normalizeSource } = await import('./source-VFLXLOCN.cjs');
  const source = await normalizeSource(ctx.source);
  const fps = ctx.videoTracks[0]?.fps ?? 30;
  const audio = new AudioOutput();
  const renderer = new VideoRenderer(target, audio, fps);
  let handles;
  try {
    handles = await startHybridDecoder({
      source,
      filename: ctx.name ?? "input.bin",
      context: ctx,
      renderer,
      audio,
      transport
    });
  } catch (err) {
    audio.destroy();
    renderer.destroy();
    throw err;
  }
  Object.defineProperty(target, "currentTime", {
    configurable: true,
    get: () => audio.now(),
    set: (v) => {
      void doSeek(v);
    }
  });
  Object.defineProperty(target, "paused", {
    configurable: true,
    get: () => !audio.isPlaying()
  });
  Object.defineProperty(target, "volume", {
    configurable: true,
    get: () => audio.getVolume(),
    set: (v) => {
      audio.setVolume(v);
      target.dispatchEvent(new Event("volumechange"));
    }
  });
  Object.defineProperty(target, "muted", {
    configurable: true,
    get: () => audio.getMuted(),
    set: (m) => {
      audio.setMuted(m);
      target.dispatchEvent(new Event("volumechange"));
    }
  });
  if (ctx.duration && Number.isFinite(ctx.duration)) {
    Object.defineProperty(target, "duration", {
      configurable: true,
      get: () => ctx.duration ?? NaN
    });
  }
  Object.defineProperty(target, "playbackRate", {
    configurable: true,
    get: () => audio.getPlaybackRate(),
    set: (v) => {
      audio.setPlaybackRate(v);
      target.dispatchEvent(new Event("ratechange"));
    }
  });
  Object.defineProperty(target, "readyState", {
    configurable: true,
    get: () => {
      if (!renderer.hasFrames()) return 0;
      if (!audio.isPlaying() && audio.bufferAhead() <= 0 && !audio.isNoAudio()) return 1;
      return 2;
    }
  });
  Object.defineProperty(target, "seekable", {
    configurable: true,
    get: () => makeTimeRanges(ctx.duration && Number.isFinite(ctx.duration) && ctx.duration > 0 ? [[0, ctx.duration]] : [])
  });
  Object.defineProperty(target, "buffered", {
    configurable: true,
    get: () => {
      const end = handles.bufferedUntilSec();
      return makeTimeRanges(end > 0 ? [[0, end]] : []);
    }
  });
  async function waitForBuffer() {
    const start = performance.now();
    while (true) {
      const audioReady = audio.isNoAudio() || audio.bufferAhead() >= READY_AUDIO_BUFFER_SECONDS;
      if (audioReady && renderer.hasFrames()) {
        return;
      }
      if ((performance.now() - start) / 1e3 > READY_TIMEOUT_SECONDS) return;
      await new Promise((r) => setTimeout(r, 50));
    }
  }
  async function doSeek(timeSec) {
    const wasPlaying = audio.isPlaying();
    target.dispatchEvent(new Event("seeking"));
    await audio.pause().catch(() => {
    });
    await handles.seek(timeSec).catch(
      (err) => console.warn("[avbridge] hybrid decoder seek failed:", err)
    );
    await audio.reset(timeSec);
    renderer.flush();
    if (wasPlaying) {
      await waitForBuffer();
      await audio.start();
    }
    target.dispatchEvent(new Event("seeked"));
  }
  queueMicrotask(() => {
    try {
      target.dispatchEvent(new Event("loadedmetadata"));
    } catch {
    }
  });
  let fatalErrorHandler = null;
  handles.onFatalError((reason) => fatalErrorHandler?.(reason));
  return {
    strategy: "hybrid",
    async play() {
      if (!audio.isPlaying()) {
        await waitForBuffer();
        await audio.start();
        target.dispatchEvent(new Event("play"));
        target.dispatchEvent(new Event("playing"));
      }
    },
    pause() {
      void audio.pause();
      target.dispatchEvent(new Event("pause"));
    },
    async seek(time) {
      await doSeek(time);
    },
    async setAudioTrack(id) {
      if (!ctx.audioTracks.some((t) => t.id === id)) {
        console.warn("[avbridge] hybrid: setAudioTrack \u2014 unknown track id", id);
        return;
      }
      const wasPlaying = audio.isPlaying();
      const currentTime = audio.now();
      await audio.pause().catch(() => {
      });
      await handles.setAudioTrack(id, currentTime).catch(
        (err) => console.warn("[avbridge] hybrid: handles.setAudioTrack failed:", err)
      );
      await audio.reset(currentTime);
      renderer.flush();
      if (wasPlaying) {
        await waitForBuffer();
        await audio.start();
      }
    },
    async setSubtitleTrack(_id) {
    },
    getCurrentTime() {
      return audio.now();
    },
    onFatalError(handler) {
      fatalErrorHandler = handler;
    },
    async destroy() {
      await handles.destroy();
      renderer.destroy();
      audio.destroy();
      try {
        delete target.currentTime;
        delete target.duration;
        delete target.paused;
        delete target.volume;
        delete target.muted;
        delete target.readyState;
        delete target.seekable;
        delete target.playbackRate;
      } catch {
      }
    },
    getRuntimeStats() {
      return handles.stats();
    }
  };
}

// src/strategies/fallback/decoder.ts
function isDebug3() {
  return typeof globalThis !== "undefined" && !!globalThis.AVBRIDGE_DEBUG;
}
async function startDecoder(opts) {
  const variant = "avbridge";
  const libav = await chunkG4APZMCP_cjs.loadLibav(variant);
  const bridge = await loadBridge2();
  const { prepareLibavInput } = await import('./libav-http-reader-Q356EO2K.cjs');
  const inputHandle = await prepareLibavInput(libav, opts.filename, opts.source, opts.transport);
  const readPkt = await libav.av_packet_alloc();
  const [fmt_ctx, streams] = await libav.ff_init_demuxer_file(opts.filename);
  const videoStream = streams.find((s) => s.codec_type === libav.AVMEDIA_TYPE_VIDEO) ?? null;
  const firstAudioTrackId = opts.context.audioTracks[0]?.id;
  let audioStream = (firstAudioTrackId != null ? streams.find((s) => s.codec_type === libav.AVMEDIA_TYPE_AUDIO && s.index === firstAudioTrackId) : void 0) ?? streams.find((s) => s.codec_type === libav.AVMEDIA_TYPE_AUDIO) ?? null;
  if (!videoStream && !audioStream) {
    throw new Error("fallback decoder: file has no decodable streams");
  }
  let videoDec = null;
  let audioDec = null;
  let videoTimeBase;
  let audioTimeBase;
  if (videoStream) {
    try {
      const [, c, pkt, frame] = await libav.ff_init_decoder(videoStream.codec_id, {
        codecpar: videoStream.codecpar
      });
      videoDec = { c, pkt, frame };
      if (videoStream.time_base_num && videoStream.time_base_den) {
        videoTimeBase = [videoStream.time_base_num, videoStream.time_base_den];
      }
    } catch (err) {
      console.error("[avbridge] failed to init video decoder:", err);
    }
  }
  if (audioStream) {
    try {
      const [, c, pkt, frame] = await libav.ff_init_decoder(audioStream.codec_id, {
        codecpar: audioStream.codecpar
      });
      audioDec = { c, pkt, frame };
      if (audioStream.time_base_num && audioStream.time_base_den) {
        audioTimeBase = [audioStream.time_base_num, audioStream.time_base_den];
      }
    } catch (err) {
      console.warn(
        "[avbridge] fallback: audio decoder unavailable \u2014 playing video with wall-clock timing:",
        err.message
      );
    }
  }
  if (!audioDec) {
    opts.audio.setNoAudio();
  }
  if (!videoDec && !audioDec) {
    await inputHandle.detach().catch(() => {
    });
    const codecs = [
      videoStream ? `video: ${opts.context.videoTracks[0]?.codec ?? "unknown"}` : null,
      audioStream ? `audio: ${opts.context.audioTracks[0]?.codec ?? "unknown"}` : null
    ].filter(Boolean).join(", ");
    throw new Error(
      `fallback decoder: could not initialize any libav decoders (${codecs}). The "${variant}" libav variant lacks software decoders for these codecs \u2014 rebuild with scripts/build-libav.sh including the missing decoder, or use a lighter strategy (native, remux, hybrid) instead.`
    );
  }
  let bsfCtx = null;
  let bsfPkt = null;
  let bsfRequiredButMissing = false;
  if (videoStream && opts.context.videoTracks[0]?.codec === "mpeg4") {
    try {
      bsfCtx = await libav.av_bsf_list_parse_str_js("mpeg4_unpack_bframes");
      if (bsfCtx != null && bsfCtx >= 0) {
        const parIn = await libav.AVBSFContext_par_in(bsfCtx);
        await libav.avcodec_parameters_copy(parIn, videoStream.codecpar);
        await libav.av_bsf_init(bsfCtx);
        bsfPkt = await libav.av_packet_alloc();
        chunkG4APZMCP_cjs.dbg.info("bsf", "mpeg4_unpack_bframes BSF active");
      } else {
        bsfRequiredButMissing = true;
        bsfCtx = null;
      }
    } catch (err) {
      bsfRequiredButMissing = true;
      bsfCtx = null;
      bsfPkt = null;
      chunkG4APZMCP_cjs.dbg.warn("bsf", `mpeg4_unpack_bframes BSF init failed: ${err.message}`);
    }
    if (bsfRequiredButMissing) {
      console.error(
        "[avbridge] MPEG-4 Part 2 (DivX/Xvid) detected but mpeg4_unpack_bframes BSF is unavailable in this libav variant. Files with packed B-frames will play with incorrect frame ordering (backwards PTS jumps, heavy late-drop stuttering). Rebuild the libav variant with the `avbsf` fragment included. See docs/dev/POSTMORTEMS.md for details."
      );
    }
  }
  async function applyBSF(packets) {
    if (!bsfCtx || !bsfPkt) return packets;
    const out = [];
    for (const pkt of packets) {
      await libav.ff_copyin_packet(bsfPkt, pkt);
      const sendErr = await libav.av_bsf_send_packet(bsfCtx, bsfPkt);
      if (sendErr < 0) {
        continue;
      }
      while (true) {
        const recvErr = await libav.av_bsf_receive_packet(bsfCtx, bsfPkt);
        if (recvErr < 0) break;
        out.push(await libav.ff_copyout_packet(bsfPkt));
      }
    }
    return out;
  }
  async function flushBSF() {
    if (!bsfCtx || !bsfPkt) return;
    try {
      if (libav.av_bsf_flush) {
        await libav.av_bsf_flush(bsfCtx);
      } else {
        while (true) {
          const err = await libav.av_bsf_receive_packet(bsfCtx, bsfPkt);
          if (err < 0) break;
        }
      }
    } catch {
    }
  }
  let destroyed = false;
  let pumpToken = 0;
  let pumpRunning = null;
  let packetsRead = 0;
  let videoFramesDecoded = 0;
  let bufferedUntilSec = 0;
  let audioFramesDecoded = 0;
  let watchdogFirstFrameMs = 0;
  let watchdogSlowSinceMs = 0;
  let watchdogSlowWarned = false;
  let watchdogOverflowWarned = false;
  let lastContentUs = -1;
  let firstValidPtsLoggedSinceSeek = false;
  let seenFirstAudioPacketSinceSeek = false;
  let seekTargetSec = 0;
  let diagPktsLoggedSinceSeek = 0;
  let diagFramesLoggedSinceSeek = 0;
  let diagFrameKeysDumped = false;
  const DIAG_MAX_PKTS = 100;
  const DIAG_MAX_FRAMES = 300;
  let videoDecodeMsTotal = 0;
  let audioDecodeMsTotal = 0;
  let videoDecodeBatches = 0;
  let audioDecodeBatches = 0;
  let readMsTotal = 0;
  let readBatches = 0;
  let pumpThrottleMsTotal = 0;
  let pumpThrottleEntries = 0;
  let slowestVideoBatchMs = 0;
  let newestVideoPtsUs = 0;
  let lastEmittedPtsUs = -1;
  let ptsRegressions = 0;
  let worstPtsRegressionMs = 0;
  const videoTrackInfo = opts.context.videoTracks.find((t) => t.id === videoStream?.index);
  const videoFps = videoTrackInfo?.fps && videoTrackInfo.fps > 0 ? videoTrackInfo.fps : 30;
  const videoFrameStepUs = Math.max(1, Math.round(1e6 / videoFps));
  async function pumpLoop(myToken) {
    while (!destroyed && myToken === pumpToken) {
      let readErr;
      let packets;
      try {
        const _readStart = performance.now();
        [readErr, packets] = await libav.ff_read_frame_multi(fmt_ctx, readPkt, {
          limit: 16 * 1024
        });
        readMsTotal += performance.now() - _readStart;
        readBatches++;
      } catch (err) {
        console.error("[avbridge] ff_read_frame_multi failed:", err);
        return;
      }
      if (myToken !== pumpToken || destroyed) return;
      const videoPackets = videoStream ? packets[videoStream.index] : void 0;
      const audioPackets = audioStream ? packets[audioStream.index] : void 0;
      if (videoPackets && videoTimeBase) {
        for (const pkt of videoPackets) {
          const sec = chunkHZUVMXBN_cjs.packetPtsSec(pkt, videoTimeBase);
          if (sec != null && sec > bufferedUntilSec) bufferedUntilSec = sec;
          if (isDebug3() && diagPktsLoggedSinceSeek < DIAG_MAX_PKTS) {
            const rawHi = pkt.ptshi ?? 0;
            const rawLo = pkt.pts ?? 0;
            const isInvalidPts = rawHi === -2147483648 && rawLo === 0;
            const rawPts64 = isInvalidPts ? null : rawHi * 4294967296 + rawLo;
            const rawSec = rawPts64 != null && videoTimeBase ? rawPts64 * videoTimeBase[0] / videoTimeBase[1] : null;
            const pktKeys = diagPktsLoggedSinceSeek === 0 ? `[keys: ${Object.keys(pkt).join(",")}]` : "";
            console.log(
              `[DIAG-PKT] vidx=${diagPktsLoggedSinceSeek} pts=${isInvalidPts ? "NOPTS" : rawPts64} pts_sec=${rawSec != null ? rawSec.toFixed(3) : "n/a"} ptshi=${rawHi} ptslo=${rawLo} flags=0x${(pkt.flags ?? 0).toString(16)} keyframe=${(pkt.flags ?? 0) & 1 ? "Y" : "N"} stream=${pkt.stream_index} dataLen=${pkt.data?.length ?? 0} seekTarget=${seekTargetSec.toFixed(3)} ` + pktKeys
            );
            diagPktsLoggedSinceSeek++;
          }
        }
      }
      if (audioPackets && audioTimeBase) {
        for (const pkt of audioPackets) {
          const sec = chunkHZUVMXBN_cjs.packetPtsSec(pkt, audioTimeBase);
          if (sec != null && sec > bufferedUntilSec) bufferedUntilSec = sec;
        }
        if (!seenFirstAudioPacketSinceSeek && audioPackets.length > 0) {
          const firstSec = chunkHZUVMXBN_cjs.packetPtsSec(audioPackets[0], audioTimeBase);
          if (firstSec != null && Number.isFinite(firstSec)) {
            seenFirstAudioPacketSinceSeek = true;
            chunkG4APZMCP_cjs.dbg.info(
              "av-anchor",
              `seek-target=${seekTargetSec.toFixed(3)}s, first-audio-pkt-pts=${firstSec.toFixed(3)}s (\u0394=${((firstSec - seekTargetSec) * 1e3).toFixed(1)}ms \u2014 pre-target packets will be skipped by AudioOutput)`
            );
          }
        }
      }
      if (audioDec && audioPackets && audioPackets.length > 0) {
        await decodeAudioBatch(
          audioPackets,
          myToken,
          /*flush*/
          false,
          audioTimeBase
        );
      }
      if (myToken !== pumpToken || destroyed) return;
      if (videoDec && videoPackets && videoPackets.length > 0) {
        const processed = await applyBSF(videoPackets);
        await decodeVideoBatch(processed, myToken);
      }
      packetsRead += (videoPackets?.length ?? 0) + (audioPackets?.length ?? 0);
      if (videoFramesDecoded > 0) {
        if (watchdogFirstFrameMs === 0) {
          watchdogFirstFrameMs = performance.now();
        }
        const elapsedSinceFirst = (performance.now() - watchdogFirstFrameMs) / 1e3;
        if (elapsedSinceFirst > 1 && !watchdogSlowWarned) {
          const expectedFrames = elapsedSinceFirst * videoFps;
          const ratio = videoFramesDecoded / expectedFrames;
          if (ratio < 0.6) {
            if (watchdogSlowSinceMs === 0) watchdogSlowSinceMs = performance.now();
            if ((performance.now() - watchdogSlowSinceMs) / 1e3 > 5) {
              watchdogSlowWarned = true;
              console.warn(
                "[avbridge:decode-rate]",
                `decoder is running slower than realtime: ${videoFramesDecoded} frames in ${elapsedSinceFirst.toFixed(1)}s (${(videoFramesDecoded / elapsedSinceFirst).toFixed(1)} fps vs ${videoFps} fps source \u2014 ${(ratio * 100).toFixed(0)}% of realtime). Playback will stutter. Typical causes: software decode of a codec with no WebCodecs support (rv40, mpeg4 @ 720p+, wmv3), or a resolution too large for single-threaded WASM to keep up with.`
              );
            }
          } else {
            watchdogSlowSinceMs = 0;
          }
        }
        if (!watchdogOverflowWarned && videoFramesDecoded > 100) {
          const rendererStats = opts.renderer.stats();
          const overflow = rendererStats.framesDroppedOverflow ?? 0;
          if (overflow / videoFramesDecoded > 0.1) {
            watchdogOverflowWarned = true;
            console.warn(
              "[avbridge:overflow-drop]",
              `renderer is dropping ${overflow}/${videoFramesDecoded} frames (${(overflow / videoFramesDecoded * 100).toFixed(0)}%) because the decoder is producing bursts faster than the canvas can drain. Symptom: choppy playback despite decoder keeping up on average. Fix would be smaller read batches in the pump loop or a lower queueHighWater cap \u2014 see src/strategies/fallback/decoder.ts.`
            );
          }
        }
      }
      {
        const _throttleStart = performance.now();
        let _throttled = false;
        while (!destroyed && myToken === pumpToken && opts.audio.bufferAhead() > 2) {
          _throttled = true;
          await new Promise((r) => setTimeout(r, 50));
        }
        if (_throttled) {
          pumpThrottleMsTotal += performance.now() - _throttleStart;
          pumpThrottleEntries++;
        }
      }
      if (readErr === libav.AVERROR_EOF) {
        if (videoDec) await decodeVideoBatch(
          [],
          myToken,
          /*flush*/
          true
        );
        if (audioDec) await decodeAudioBatch(
          [],
          myToken,
          /*flush*/
          true
        );
        return;
      }
      if (readErr && readErr !== 0 && readErr !== -libav.EAGAIN) {
        console.warn("[avbridge] ff_read_frame_multi returned", readErr);
        return;
      }
    }
  }
  async function decodeVideoBatch(pkts, myToken, flush = false) {
    if (!videoDec || destroyed || myToken !== pumpToken) return;
    let frames;
    const _t0 = performance.now();
    try {
      frames = await libav.ff_decode_multi(
        videoDec.c,
        videoDec.pkt,
        videoDec.frame,
        pkts,
        flush ? { fin: true, ignoreErrors: true } : { ignoreErrors: true }
      );
    } catch (err) {
      console.error("[avbridge] video decode batch failed:", err);
      return;
    }
    {
      const _dt = performance.now() - _t0;
      videoDecodeMsTotal += _dt;
      videoDecodeBatches++;
      if (_dt > slowestVideoBatchMs) slowestVideoBatchMs = _dt;
    }
    if (myToken !== pumpToken || destroyed) return;
    for (const f of frames) {
      if (myToken !== pumpToken || destroyed) return;
      const _diagShouldLog = isDebug3() && diagFramesLoggedSinceSeek < DIAG_MAX_FRAMES;
      const _diagRawHi = f.ptshi ?? 0;
      const _diagRawLo = f.pts ?? 0;
      const _diagInvalid = _diagRawHi === -2147483648 && _diagRawLo === 0;
      const _diagRawPts64 = _diagInvalid ? null : _diagRawHi * 4294967296 + _diagRawLo;
      const _diagRawSec = _diagRawPts64 != null && videoTimeBase ? _diagRawPts64 * videoTimeBase[0] / videoTimeBase[1] : null;
      if (_diagShouldLog && !diagFrameKeysDumped) {
        diagFrameKeysDumped = true;
        const allKeys = Object.keys(f);
        const fieldDump = {};
        for (const k of allKeys) {
          const v = f[k];
          if (k === "data") continue;
          if (typeof v === "object" && v !== null && "length" in v) continue;
          fieldDump[k] = v;
        }
        console.log(`[DIAG-FRAME] FIRST FRAME post-seek \u2014 all keys: ${allKeys.join(",")}`);
        console.log(`[DIAG-FRAME] FIRST FRAME field dump:`, fieldDump);
      }
      let rawUs = null;
      if (!_diagInvalid && _diagRawPts64 != null) {
        const tb = videoTimeBase ?? [1, 1e6];
        const us = Math.round(_diagRawPts64 * 1e6 * tb[0] / tb[1]);
        if (Number.isFinite(us) && Math.abs(us) <= Number.MAX_SAFE_INTEGER) {
          rawUs = us;
        }
      }
      const _diagLog = (decision, finalPtsUs, sanFallback) => {
        if (!_diagShouldLog) return;
        const ptsSrc = sanFallback ? `SYNTHETIC(${_diagInvalid ? "NOPTS" : "invalid-range"})` : "LIBAV";
        console.log(
          `[DIAG-FRAME] vidx=${diagFramesLoggedSinceSeek} raw_pts=${_diagInvalid ? "NOPTS" : _diagRawPts64} raw_pts_sec=${_diagRawSec != null ? _diagRawSec.toFixed(3) : "n/a"} pts_src=${ptsSrc} final_pts_us=${finalPtsUs} final_pts_sec=${(finalPtsUs / 1e6).toFixed(3)} seekTarget=${seekTargetSec.toFixed(3)} offset_to_target_ms=${(finalPtsUs / 1e3 - seekTargetSec * 1e3).toFixed(1)} lastEmittedPts_us=${lastEmittedPtsUs} decision=${decision}`
        );
        diagFramesLoggedSinceSeek++;
      };
      let _diagSanFallbackFired = false;
      const seekTargetUs = Math.round(seekTargetSec * 1e6);
      if (lastContentUs < 0) {
        if (rawUs == null) {
          const isColdStartKeyframe = seekTargetSec === 0 && f.key_frame === 1;
          if (isColdStartKeyframe) {
            lastContentUs = 0;
            _diagSanFallbackFired = true;
          } else {
            _diagLog("PRE-ANCHOR-DROP", 0, true);
            continue;
          }
        } else {
          lastContentUs = rawUs;
          if (!firstValidPtsLoggedSinceSeek) {
            firstValidPtsLoggedSinceSeek = true;
            if (isDebug3()) {
              console.log(
                `[avbridge:decoder] post-seek anchor established: first valid raw pts = ${(rawUs / 1e3).toFixed(1)}ms (seekTarget = ${(seekTargetSec * 1e3).toFixed(1)}ms, \u0394 = ${((rawUs - seekTargetUs) / 1e3).toFixed(1)}ms)`
              );
            }
            if (rawUs >= seekTargetUs) {
              console.warn(
                `[avbridge:decoder] first valid raw pts \u2265 seek target \u2014 pre-anchor NOPTS frames may have straddled the target and been mis-discarded. First painted frame may be late by up to one keyframe interval.`
              );
            }
          }
        }
      } else {
        if (rawUs != null) {
          lastContentUs = rawUs;
        } else {
          lastContentUs += videoFrameStepUs;
          _diagSanFallbackFired = true;
        }
      }
      f.pts = lastContentUs;
      f.ptshi = lastContentUs < 0 ? -1 : 0;
      const _fPts = lastContentUs;
      if (_fPts > newestVideoPtsUs) newestVideoPtsUs = _fPts;
      if (lastEmittedPtsUs >= 0 && _fPts < lastEmittedPtsUs) {
        _diagLog("REGRESSED-DROP", _fPts, _diagSanFallbackFired);
        ptsRegressions++;
        const regressMs = (lastEmittedPtsUs - _fPts) / 1e3;
        if (regressMs > worstPtsRegressionMs) worstPtsRegressionMs = regressMs;
        if (ptsRegressions <= 10) {
          console.warn(
            `[avbridge:decoder] dropped out-of-order frame #${ptsRegressions}: pts=${(_fPts / 1e3).toFixed(1)}ms < previous=${(lastEmittedPtsUs / 1e3).toFixed(1)}ms (regression=${regressMs.toFixed(1)}ms). Typically a post-seek B-frame reorder tail.`
          );
        }
        continue;
      }
      lastEmittedPtsUs = _fPts;
      const targetUs = Math.round(seekTargetSec * 1e6);
      if (_fPts < targetUs - videoFrameStepUs) {
        _diagLog("PRE-TARGET-DROP", _fPts, _diagSanFallbackFired);
        continue;
      }
      try {
        const vf = bridge.laFrameToVideoFrame(f, { timeBase: [1, 1e6] });
        if (opts.renderer.queueDepth() >= opts.renderer.queueHighWater) {
          vf.close();
          _diagLog("OVERFLOW-DROP", _fPts, _diagSanFallbackFired);
        } else {
          opts.renderer.enqueue(vf);
          _diagLog("ENQUEUED", _fPts, _diagSanFallbackFired);
        }
        videoFramesDecoded++;
      } catch (err) {
        if (videoFramesDecoded === 0) {
          console.warn("[avbridge] laFrameToVideoFrame failed:", err);
        }
        _diagLog("BRIDGE-ERROR", _fPts, _diagSanFallbackFired);
      }
    }
  }
  async function decodeAudioBatch(pkts, myToken, flush = false, tb) {
    if (!audioDec || destroyed || myToken !== pumpToken) return;
    const pktPtsSec = pkts.map(
      (p) => tb ? chunkHZUVMXBN_cjs.packetPtsSec(p, tb) : null
    );
    let frames;
    const _t0 = performance.now();
    try {
      frames = await libav.ff_decode_multi(
        audioDec.c,
        audioDec.pkt,
        audioDec.frame,
        pkts,
        flush ? { fin: true, ignoreErrors: true } : { ignoreErrors: true }
      );
    } catch (err) {
      console.error("[avbridge] audio decode batch failed:", err);
      return;
    }
    audioDecodeMsTotal += performance.now() - _t0;
    audioDecodeBatches++;
    if (myToken !== pumpToken || destroyed) return;
    for (let i = 0; i < frames.length; i++) {
      if (myToken !== pumpToken || destroyed) return;
      const f = frames[i];
      const samples = chunkHZUVMXBN_cjs.libavFrameToInterleavedFloat32(f);
      if (samples) {
        const pts = pktPtsSec[i] ?? null;
        if (isDebug3()) {
          const dur = samples.data.length / samples.channels / samples.sampleRate;
          console.log(`[TRACE-DEC] audio frame #${audioFramesDecoded} pts=${pts != null ? pts.toFixed(4) : "NULL"} dur=${dur.toFixed(4)} samples=${samples.data.length / samples.channels} sr=${samples.sampleRate} ch=${samples.channels} pktsIn=${pkts.length} framesOut=${frames.length}`);
        }
        opts.audio.schedule(samples.data, samples.channels, samples.sampleRate, pts);
        audioFramesDecoded++;
      }
    }
  }
  pumpToken = 1;
  pumpRunning = pumpLoop(pumpToken).catch(
    (err) => console.error("[avbridge] decoder pump failed:", err)
  );
  return {
    async destroy() {
      destroyed = true;
      pumpToken++;
      try {
        await pumpRunning;
      } catch {
      }
      try {
        if (bsfCtx) await libav.av_bsf_free(bsfCtx);
      } catch {
      }
      try {
        if (bsfPkt) await libav.av_packet_free?.(bsfPkt);
      } catch {
      }
      try {
        if (videoDec) await libav.ff_free_decoder?.(videoDec.c, videoDec.pkt, videoDec.frame);
      } catch {
      }
      try {
        if (audioDec) await libav.ff_free_decoder?.(audioDec.c, audioDec.pkt, audioDec.frame);
      } catch {
      }
      try {
        await libav.av_packet_free?.(readPkt);
      } catch {
      }
      try {
        await libav.avformat_close_input_js(fmt_ctx);
      } catch {
      }
      try {
        await inputHandle.detach();
      } catch {
      }
    },
    async setAudioTrack(trackId, timeSec) {
      if (audioStream && audioStream.index === trackId) return;
      const newStream = streams.find(
        (s) => s.codec_type === libav.AVMEDIA_TYPE_AUDIO && s.index === trackId
      );
      if (!newStream) {
        console.warn("[avbridge] fallback: setAudioTrack \u2014 no stream with id", trackId);
        return;
      }
      const newToken = ++pumpToken;
      if (pumpRunning) {
        try {
          await pumpRunning;
        } catch {
        }
      }
      if (destroyed) return;
      if (audioDec) {
        try {
          await libav.ff_free_decoder?.(audioDec.c, audioDec.pkt, audioDec.frame);
        } catch {
        }
        audioDec = null;
      }
      try {
        const [, c, pkt, frame] = await libav.ff_init_decoder(newStream.codec_id, {
          codecpar: newStream.codecpar
        });
        audioDec = { c, pkt, frame };
        audioTimeBase = newStream.time_base_num && newStream.time_base_den ? [newStream.time_base_num, newStream.time_base_den] : void 0;
      } catch (err) {
        console.warn(
          "[avbridge] fallback: setAudioTrack init failed \u2014 falling back to no-audio mode:",
          err.message
        );
        audioDec = null;
        opts.audio.setNoAudio();
      }
      audioStream = newStream;
      try {
        const tsUs = Math.floor(timeSec * 1e6);
        const [tsLo, tsHi] = libav.f64toi64 ? libav.f64toi64(tsUs) : [tsUs | 0, Math.floor(tsUs / 4294967296)];
        await libav.av_seek_frame(
          fmt_ctx,
          -1,
          tsLo,
          tsHi,
          libav.AVSEEK_FLAG_BACKWARD ?? 0
        );
      } catch (err) {
        console.warn("[avbridge] fallback: setAudioTrack seek failed:", err);
      }
      try {
        if (videoDec) await libav.avcodec_flush_buffers?.(videoDec.c);
      } catch {
      }
      await flushBSF();
      lastContentUs = -1;
      lastEmittedPtsUs = -1;
      firstValidPtsLoggedSinceSeek = false;
      pumpRunning = pumpLoop(newToken).catch(
        (err) => console.error("[avbridge] fallback pump failed (post-setAudioTrack):", err)
      );
    },
    async seek(timeSec) {
      if (isDebug3()) {
        console.log(`[SEEK] target=${timeSec.toFixed(3)}s (${(timeSec * 1e3).toFixed(0)}ms) wall=${performance.now().toFixed(0)}`);
      }
      const newToken = ++pumpToken;
      if (pumpRunning) {
        try {
          await pumpRunning;
        } catch {
        }
      }
      if (destroyed) return;
      try {
        const tsUs = Math.floor(timeSec * 1e6);
        const [tsLo, tsHi] = libav.f64toi64 ? libav.f64toi64(tsUs) : [tsUs | 0, Math.floor(tsUs / 4294967296)];
        await libav.av_seek_frame(
          fmt_ctx,
          -1,
          tsLo,
          tsHi,
          libav.AVSEEK_FLAG_BACKWARD ?? 0
        );
      } catch (err) {
        console.warn("[avbridge] av_seek_frame failed:", err);
      }
      try {
        if (videoDec) await libav.avcodec_flush_buffers?.(videoDec.c);
      } catch {
      }
      try {
        if (audioDec) await libav.avcodec_flush_buffers?.(audioDec.c);
      } catch {
      }
      await flushBSF();
      lastContentUs = -1;
      lastEmittedPtsUs = -1;
      firstValidPtsLoggedSinceSeek = false;
      seenFirstAudioPacketSinceSeek = false;
      seekTargetSec = timeSec;
      diagPktsLoggedSinceSeek = 0;
      diagFramesLoggedSinceSeek = 0;
      diagFrameKeysDumped = false;
      pumpRunning = pumpLoop(newToken).catch(
        (err) => console.error("[avbridge] decoder pump failed (post-seek):", err)
      );
    },
    bufferedUntilSec() {
      return bufferedUntilSec;
    },
    stats() {
      return {
        decoderType: "libav-wasm",
        packetsRead,
        videoFramesDecoded,
        audioFramesDecoded,
        // Throughput instrumentation — the stats panel turns these into
        // "decode fps actual / realtime target" and shows slowest batch
        // + producer throttle share.
        videoDecodeMsTotal,
        videoDecodeBatches,
        audioDecodeMsTotal,
        audioDecodeBatches,
        readMsTotal,
        readBatches,
        pumpThrottleMsTotal,
        pumpThrottleEntries,
        slowestVideoBatchMs,
        newestVideoPtsMs: Math.round(newestVideoPtsUs / 1e3),
        ptsRegressions,
        worstPtsRegressionMs,
        sourceFps: videoFps,
        bsfApplied: bsfCtx ? ["mpeg4_unpack_bframes"] : [],
        bsfMissing: bsfRequiredButMissing ? ["mpeg4_unpack_bframes"] : [],
        // Confirmed transport info: once prepareLibavInput returns
        // successfully, we *know* whether the source is http-range (probe
        // succeeded and returned 206) or in-memory blob. Diagnostics hoists
        // these `_`-prefixed keys to the typed fields.
        _transport: inputHandle.transport === "http-range" ? "http-range" : "memory",
        _rangeSupported: inputHandle.transport === "http-range",
        ...opts.renderer.stats(),
        ...opts.audio.stats()
      };
    }
  };
}
async function loadBridge2() {
  try {
    const wrapper = await import('./libav-import-2ZVKV2E7.cjs');
    return wrapper.libavBridge;
  } catch (err) {
    throw new Error(
      `failed to load libavjs-webcodecs-bridge \u2014 install the optional peer deps with: npm i libavjs-webcodecs-bridge @libav.js/variant-webcodecs. (${err.message})`
    );
  }
}

// src/strategies/fallback/index.ts
var READY_AUDIO_BUFFER_SECONDS2 = 0.04;
var READY_TIMEOUT_SECONDS2 = 3;
async function createFallbackSession(ctx, target, transport) {
  const { normalizeSource } = await import('./source-VFLXLOCN.cjs');
  const source = await normalizeSource(ctx.source);
  const fps = ctx.videoTracks[0]?.fps ?? 30;
  const audio = new AudioOutput();
  const renderer = new VideoRenderer(target, audio, fps);
  let handles;
  try {
    handles = await startDecoder({
      source,
      filename: ctx.name ?? "input.bin",
      context: ctx,
      renderer,
      audio,
      transport
    });
  } catch (err) {
    audio.destroy();
    renderer.destroy();
    throw err;
  }
  Object.defineProperty(target, "currentTime", {
    configurable: true,
    get: () => audio.now(),
    set: (v) => {
      void doSeek(v);
    }
  });
  Object.defineProperty(target, "paused", {
    configurable: true,
    get: () => !audio.isPlaying()
  });
  Object.defineProperty(target, "volume", {
    configurable: true,
    get: () => audio.getVolume(),
    set: (v) => {
      audio.setVolume(v);
      target.dispatchEvent(new Event("volumechange"));
    }
  });
  Object.defineProperty(target, "muted", {
    configurable: true,
    get: () => audio.getMuted(),
    set: (m) => {
      audio.setMuted(m);
      target.dispatchEvent(new Event("volumechange"));
    }
  });
  if (ctx.duration && Number.isFinite(ctx.duration)) {
    Object.defineProperty(target, "duration", {
      configurable: true,
      get: () => ctx.duration ?? NaN
    });
  }
  Object.defineProperty(target, "playbackRate", {
    configurable: true,
    get: () => audio.getPlaybackRate(),
    set: (v) => {
      audio.setPlaybackRate(v);
      target.dispatchEvent(new Event("ratechange"));
    }
  });
  Object.defineProperty(target, "readyState", {
    configurable: true,
    get: () => {
      if (!renderer.hasFrames()) return 0;
      if (!audio.isPlaying() && audio.bufferAhead() <= 0 && !audio.isNoAudio()) return 1;
      return 2;
    }
  });
  Object.defineProperty(target, "seekable", {
    configurable: true,
    get: () => makeTimeRanges(ctx.duration && Number.isFinite(ctx.duration) && ctx.duration > 0 ? [[0, ctx.duration]] : [])
  });
  Object.defineProperty(target, "buffered", {
    configurable: true,
    get: () => {
      const end = handles.bufferedUntilSec();
      return makeTimeRanges(end > 0 ? [[0, end]] : []);
    }
  });
  async function waitForBuffer() {
    const start = performance.now();
    let firstFrameAtMs = 0;
    chunkG4APZMCP_cjs.dbg.info(
      "cold-start",
      `gate entry: want audio \u2265 ${READY_AUDIO_BUFFER_SECONDS2 * 1e3}ms + 1 frame`
    );
    while (true) {
      const audioAhead = audio.isNoAudio() ? Infinity : audio.bufferAhead();
      const audioReady = audio.isNoAudio() || audioAhead >= READY_AUDIO_BUFFER_SECONDS2;
      const hasFrames = renderer.hasFrames();
      const nowMs = performance.now();
      if (hasFrames && firstFrameAtMs === 0) firstFrameAtMs = nowMs;
      if (audioReady && hasFrames) {
        chunkG4APZMCP_cjs.dbg.info(
          "cold-start",
          `gate satisfied in ${(nowMs - start).toFixed(0)}ms (audio=${(audioAhead * 1e3).toFixed(0)}ms, frames=${renderer.queueDepth()})`
        );
        return;
      }
      if (hasFrames && firstFrameAtMs > 0 && nowMs - firstFrameAtMs >= 500) {
        chunkG4APZMCP_cjs.dbg.info(
          "cold-start",
          `gate released on video-only grace at ${(nowMs - start).toFixed(0)}ms (frames=${renderer.queueDepth()}, audio=${(audioAhead * 1e3).toFixed(0)}ms \u2014 demuxer hasn't delivered audio packets yet, starting anyway and letting the audio scheduler catch up at its media-time anchor)`
        );
        return;
      }
      if ((nowMs - start) / 1e3 > READY_TIMEOUT_SECONDS2) {
        chunkG4APZMCP_cjs.dbg.diag(
          "cold-start",
          `gate TIMEOUT after ${READY_TIMEOUT_SECONDS2}s \u2014 audio=${(audioAhead * 1e3).toFixed(0)}ms (needed ${READY_AUDIO_BUFFER_SECONDS2 * 1e3}ms), frames=${renderer.queueDepth()} (needed \u22651). Decoder produced nothing in ${READY_TIMEOUT_SECONDS2}s \u2014 either a corrupt source, a missing codec, or WASM is catastrophically slow on this file. Check getDiagnostics().runtime for decode counters.`
        );
        return;
      }
      await new Promise((r) => setTimeout(r, 50));
    }
  }
  async function doSeek(timeSec) {
    const wasPlaying = audio.isPlaying();
    target.dispatchEvent(new Event("seeking"));
    await audio.pause().catch(() => {
    });
    await handles.seek(timeSec).catch(
      (err) => console.warn("[avbridge] decoder seek failed:", err)
    );
    await audio.reset(timeSec);
    renderer.flush();
    if (wasPlaying) {
      await waitForBuffer();
      await audio.start();
    }
    target.dispatchEvent(new Event("seeked"));
  }
  queueMicrotask(() => {
    try {
      target.dispatchEvent(new Event("loadedmetadata"));
    } catch {
    }
  });
  return {
    strategy: "fallback",
    async play() {
      if (!audio.isPlaying()) {
        await waitForBuffer();
        await audio.start();
        target.dispatchEvent(new Event("play"));
        target.dispatchEvent(new Event("playing"));
      }
    },
    pause() {
      void audio.pause();
      target.dispatchEvent(new Event("pause"));
    },
    async seek(time) {
      await doSeek(time);
    },
    async setAudioTrack(id) {
      if (!ctx.audioTracks.some((t) => t.id === id)) {
        console.warn("[avbridge] fallback: setAudioTrack \u2014 unknown track id", id);
        return;
      }
      const wasPlaying = audio.isPlaying();
      const currentTime = audio.now();
      await audio.pause().catch(() => {
      });
      await handles.setAudioTrack(id, currentTime).catch(
        (err) => console.warn("[avbridge] fallback: handles.setAudioTrack failed:", err)
      );
      await audio.reset(currentTime);
      renderer.flush();
      if (wasPlaying) {
        await waitForBuffer();
        await audio.start();
      }
    },
    async setSubtitleTrack(_id) {
    },
    getCurrentTime() {
      return audio.now();
    },
    async destroy() {
      await handles.destroy();
      renderer.destroy();
      audio.destroy();
      try {
        delete target.currentTime;
        delete target.duration;
        delete target.paused;
        delete target.volume;
        delete target.muted;
        delete target.readyState;
        delete target.seekable;
        delete target.playbackRate;
      } catch {
      }
    },
    getRuntimeStats() {
      return handles.stats();
    }
  };
}

// src/plugins/builtin.ts
var nativePlugin = {
  name: "native",
  canHandle: () => true,
  execute: (ctx, video) => createNativeSession(ctx, video)
};
var remuxPlugin = {
  name: "remux",
  canHandle: () => true,
  execute: (ctx, video) => createRemuxSession(ctx, video)
};
var hybridPlugin = {
  name: "hybrid",
  canHandle: () => typeof VideoDecoder !== "undefined",
  execute: (ctx, video, transport) => createHybridSession(ctx, video, transport)
};
var fallbackPlugin = {
  name: "fallback",
  canHandle: () => true,
  execute: (ctx, video, transport) => createFallbackSession(ctx, video, transport)
};
function registerBuiltins(registry) {
  registry.register(nativePlugin);
  registry.register(remuxPlugin);
  registry.register(hybridPlugin);
  registry.register(fallbackPlugin);
}

// src/player.ts
function readDecodedFrameCount(target) {
  if (typeof HTMLVideoElement === "undefined" || !(target instanceof HTMLVideoElement)) return 0;
  const vq = target.getVideoPlaybackQuality;
  if (typeof vq === "function") {
    try {
      return vq.call(target).totalVideoFrames;
    } catch {
    }
  }
  const legacy = target.webkitDecodedFrameCount;
  return typeof legacy === "number" ? legacy : 0;
}
function evaluateDecodeHealth(input) {
  const timeThreshold = input.timeStallThresholdMs ?? 5e3;
  const frameThreshold = input.frameStallThresholdMs ?? 3e3;
  if (!input.timeAdvanced && input.now - input.lastProgressTime > timeThreshold) {
    return { escalate: true, kind: "time-stall" };
  }
  if (input.hasVideoTrack && input.timeAdvanced && !input.framesAdvanced && input.now - input.lastFrameProgressTime > frameThreshold) {
    return { escalate: true, kind: "silent-video" };
  }
  return { escalate: false };
}
var UnifiedPlayer = class _UnifiedPlayer {
  /**
   * @internal Use {@link createPlayer} or {@link UnifiedPlayer.create} instead.
   */
  constructor(options, registry) {
    this.options = options;
    this.registry = registry;
    const { requestInit, fetchFn, cacheBytes } = options;
    if (requestInit || fetchFn || cacheBytes !== void 0) {
      this.transport = { requestInit, fetchFn, cacheBytes };
    }
  }
  options;
  registry;
  emitter = new TypedEmitter();
  session = null;
  diag = new Diagnostics();
  timeupdateInterval = null;
  // Saved from bootstrap for strategy switching
  mediaContext = null;
  classification = null;
  // Stall detection
  stallTimer = null;
  lastProgressTime = 0;
  lastProgressPosition = -1;
  /** Last observed `HTMLVideoElement.getVideoPlaybackQuality().totalVideoFrames`
   *  (or `webkitDecodedFrameCount` fallback). Used by the silent-video
   *  watchdog — catches cases where `currentTime` advances (audio plays)
   *  but the decoder produces no frames, e.g. Firefox claiming `hev1.*`
   *  via MSE when the decoder actually can't decode HEVC. */
  lastVideoFrameCount = 0;
  lastVideoFrameProgressTime = 0;
  errorListener = null;
  // Bound so we can removeEventListener in destroy(); without this the
  // listener outlives the player and accumulates on elements that swap
  // source (e.g. <avbridge-video>).
  endedListener = null;
  // Background tab handling. userIntent is what the user last asked for
  // (play vs pause) — used to decide whether to auto-resume on visibility
  // return. autoPausedForVisibility tracks whether we paused because the
  // tab was hidden, so we don't resume playback the user deliberately
  // paused (e.g. via media keys while hidden).
  userIntent = "pause";
  autoPausedForVisibility = false;
  visibilityListener = null;
  // Serializes escalation / setStrategy calls
  switchingPromise = Promise.resolve();
  // Owns blob URLs created during sidecar discovery + SRT->VTT conversion.
  // Revoked at destroy() so repeated source swaps don't leak.
  subtitleResources = new chunkWRKO6Q42_cjs.SubtitleResourceBag();
  // Transport config extracted from CreatePlayerOptions. Threaded to probe,
  // subtitle fetches, and strategy session creators. Not stored on MediaContext
  // because it's runtime config, not media analysis.
  transport;
  static async create(options) {
    const registry = new PluginRegistry();
    registerBuiltins(registry);
    if (options.plugins) {
      for (const p of options.plugins) registry.register(
        p,
        /* prepend */
        true
      );
    }
    const player = new _UnifiedPlayer(options, registry);
    try {
      await player.bootstrap();
    } catch (err) {
      err.player = player;
      throw err;
    }
    return player;
  }
  async bootstrap() {
    const bootstrapStart = performance.now();
    try {
      chunkG4APZMCP_cjs.dbg.info("bootstrap", "start");
      const ctx = await chunkG4APZMCP_cjs.dbg.timed("probe", "probe", 3e3, () => chunkVLI3Y6IJ_cjs.probe(this.options.source, this.transport));
      chunkG4APZMCP_cjs.dbg.info(
        "probe",
        `container=${ctx.container} video=${ctx.videoTracks[0]?.codec ?? "-"} audio=${ctx.audioTracks[0]?.codec ?? "-"} probedBy=${ctx.probedBy}`
      );
      this.diag.recordProbe(ctx);
      this.mediaContext = ctx;
      if (this.options.subtitles) {
        for (const s of this.options.subtitles) {
          ctx.subtitleTracks.push({
            id: ctx.subtitleTracks.length,
            format: s.format ?? (s.url.endsWith(".srt") ? "srt" : "vtt"),
            language: s.language,
            sidecarUrl: s.url
          });
        }
      }
      if (this.options.directory && this.options.source instanceof File) {
        const found = await chunkWRKO6Q42_cjs.discoverSidecars(this.options.source, this.options.directory);
        for (const s of found) {
          this.subtitleResources.track(s.url);
          ctx.subtitleTracks.push({
            id: ctx.subtitleTracks.length,
            format: s.format,
            language: s.language,
            sidecarUrl: s.url
          });
        }
      }
      const decision = this.options.initialStrategy ? buildInitialDecision(this.options.initialStrategy, ctx) : classifyContext(ctx);
      chunkG4APZMCP_cjs.dbg.info(
        "classify",
        `strategy=${decision.strategy} class=${decision.class} reason="${decision.reason}"` + (decision.fallbackChain ? ` fallback=${decision.fallbackChain.join("\u2192")}` : "")
      );
      this.classification = decision;
      this.diag.recordClassification(decision);
      this.emitter.emitSticky("strategy", {
        strategy: decision.strategy,
        reason: decision.reason
      });
      await this.startSession(decision.strategy, decision.reason);
      await chunkWRKO6Q42_cjs.attachSubtitleTracks(
        this.options.target,
        ctx.subtitleTracks,
        this.subtitleResources,
        (err, track) => {
          console.warn(`[avbridge] subtitle ${track.id} failed: ${err.message}`);
        },
        this.transport
      );
      this.emitter.emitSticky("tracks", {
        video: ctx.videoTracks,
        audio: ctx.audioTracks,
        subtitle: ctx.subtitleTracks
      });
      this.startTimeupdateLoop();
      this.endedListener = () => this.emitter.emit("ended", void 0);
      this.options.target.addEventListener("ended", this.endedListener);
      if (this.options.backgroundBehavior !== "continue" && typeof document !== "undefined") {
        this.visibilityListener = () => this.onVisibilityChange();
        document.addEventListener("visibilitychange", this.visibilityListener);
      }
      this.emitter.emitSticky("ready", void 0);
      const bootstrapElapsed = performance.now() - bootstrapStart;
      chunkG4APZMCP_cjs.dbg.info("bootstrap", `ready in ${bootstrapElapsed.toFixed(0)}ms`);
      if (bootstrapElapsed > 5e3) {
        console.warn(
          "[avbridge:bootstrap]",
          `total bootstrap time ${bootstrapElapsed.toFixed(0)}ms \u2014 unusually slow. Enable globalThis.AVBRIDGE_DEBUG for a per-phase breakdown.`
        );
      }
    } catch (err) {
      const e = err instanceof Error ? err : new Error(String(err));
      this.diag.recordError(e);
      this.emitter.emit("error", e);
      throw e;
    }
  }
  /**
   * Try to start a session with the given strategy. On failure, walk the
   * fallback chain. Throws only if all strategies are exhausted.
   */
  async startSession(strategy, reason) {
    const plugin = this.registry.findFor(this.mediaContext, strategy);
    if (!plugin) {
      throw new Error(`no plugin available for strategy "${strategy}"`);
    }
    try {
      this.session = await plugin.execute(this.mediaContext, this.options.target, this.transport);
    } catch (err) {
      const chain = this.classification?.fallbackChain;
      if (chain && chain.length > 0) {
        const next = chain.shift();
        console.warn(`[avbridge] ${strategy} failed (${err.message}), escalating to ${next}`);
        this.emitter.emit("strategychange", {
          from: strategy,
          to: next,
          reason: `${strategy} failed: ${err.message}`,
          currentTime: 0
        });
        this.diag.recordStrategySwitch(next, `${strategy} failed: ${err.message}`);
        return this.startSession(next, `escalated from ${strategy}`);
      }
      throw err;
    }
    this.session.onFatalError?.((fatalReason) => {
      void this.escalate(fatalReason);
    });
    this.attachSupervisor();
    if (this.session.strategy !== strategy) {
      this.emitter.emitSticky("strategy", {
        strategy: this.session.strategy,
        reason
      });
    }
  }
  // ── Escalation ──────────────────────────────────────────────────────────
  async escalate(reason) {
    this.switchingPromise = this.switchingPromise.then(
      () => this.doEscalate(reason)
    ).catch((err) => {
      this.emitter.emit("error", err instanceof Error ? err : new Error(String(err)));
    });
    await this.switchingPromise;
  }
  async doEscalate(reason) {
    const chain = this.classification?.fallbackChain;
    if (!chain || chain.length === 0) {
      this.emitter.emit("error", new Error(
        `strategy "${this.session?.strategy}" failed: ${reason} (no fallback available)`
      ));
      return;
    }
    const currentTime = this.session?.getCurrentTime() ?? 0;
    const wasPlaying = this.session ? !this.options.target.paused : false;
    const fromStrategy = this.session?.strategy ?? "native";
    this.clearSupervisor();
    if (this.session) {
      try {
        await this.session.destroy();
      } catch {
      }
      this.session = null;
    }
    const errors = [];
    while (chain.length > 0) {
      const nextStrategy = chain.shift();
      console.warn(`[avbridge] escalating from ${fromStrategy} to ${nextStrategy}: ${reason}`);
      this.emitter.emit("strategychange", {
        from: fromStrategy,
        to: nextStrategy,
        reason,
        currentTime
      });
      this.diag.recordStrategySwitch(nextStrategy, reason);
      const plugin = this.registry.findFor(this.mediaContext, nextStrategy);
      if (!plugin) {
        errors.push(`${nextStrategy}: no plugin available`);
        continue;
      }
      try {
        this.session = await plugin.execute(this.mediaContext, this.options.target, this.transport);
      } catch (err) {
        const msg = err instanceof Error ? err.message : String(err);
        errors.push(`${nextStrategy}: ${msg}`);
        console.warn(`[avbridge] ${nextStrategy} failed during escalation, trying next: ${msg}`);
        continue;
      }
      this.emitter.emitSticky("strategy", {
        strategy: nextStrategy,
        reason: `escalated: ${reason}`
      });
      this.session.onFatalError?.((fatalReason) => {
        void this.escalate(fatalReason);
      });
      this.attachSupervisor();
      try {
        await this.session.seek(currentTime);
        if (wasPlaying) await this.session.play();
      } catch (err) {
        console.warn("[avbridge] failed to restore position after escalation:", err);
      }
      return;
    }
    this.emitter.emit("error", new chunk2IJ66NTD_cjs.AvbridgeError(
      chunk2IJ66NTD_cjs.ERR_ALL_STRATEGIES_EXHAUSTED,
      `All playback strategies failed: ${errors.join("; ")}`,
      "This file may require a codec or container that isn't available in this browser. Try the fallback strategy or check browser codec support."
    ));
  }
  // ── Stall supervision ─────────────────────────────────────────────────
  attachSupervisor() {
    this.clearSupervisor();
    if (this.options.autoEscalate === false) return;
    if (!this.classification?.fallbackChain?.length) return;
    const strategy = this.session?.strategy;
    if (strategy === "native" || strategy === "remux") {
      this.lastProgressPosition = this.options.target.currentTime;
      this.lastProgressTime = performance.now();
      this.lastVideoFrameCount = readDecodedFrameCount(this.options.target);
      this.lastVideoFrameProgressTime = performance.now();
      const hasVideoTrack = (this.mediaContext?.videoTracks.length ?? 0) > 0;
      this.stallTimer = setInterval(() => {
        const t = this.options.target;
        const now = performance.now();
        if (t.paused || t.ended || t.readyState < 2) {
          this.lastProgressPosition = t.currentTime;
          this.lastProgressTime = now;
          this.lastVideoFrameCount = readDecodedFrameCount(t);
          this.lastVideoFrameProgressTime = now;
          return;
        }
        const timeAdvanced = t.currentTime !== this.lastProgressPosition;
        const frames = readDecodedFrameCount(t);
        const framesAdvanced = frames > this.lastVideoFrameCount;
        const health = evaluateDecodeHealth({
          hasVideoTrack,
          timeAdvanced,
          framesAdvanced,
          now,
          lastProgressTime: this.lastProgressTime,
          lastFrameProgressTime: this.lastVideoFrameProgressTime
        });
        if (timeAdvanced) {
          this.lastProgressPosition = t.currentTime;
          this.lastProgressTime = now;
        }
        if (framesAdvanced) {
          this.lastVideoFrameCount = frames;
          this.lastVideoFrameProgressTime = now;
        }
        if (health.escalate) {
          const reason = health.kind === "time-stall" ? `${strategy} strategy stalled for 5s at ${t.currentTime.toFixed(1)}s` : `${strategy} strategy: audio is advancing but the video decoder has produced no new frames for 3s \u2014 likely a silent codec failure`;
          void this.escalate(reason);
        }
      }, 1e3);
      const onError = () => {
        void this.escalate(
          `${strategy} strategy error: ${this.options.target.error?.message ?? "unknown"}`
        );
      };
      this.options.target.addEventListener("error", onError, { once: true });
      this.errorListener = onError;
    }
  }
  clearSupervisor() {
    if (this.stallTimer) {
      clearInterval(this.stallTimer);
      this.stallTimer = null;
    }
    if (this.errorListener) {
      this.options.target.removeEventListener("error", this.errorListener);
      this.errorListener = null;
    }
  }
  // ── Public: manual strategy switch ────────────────────────────────────
  /** Manually switch to a different playback strategy. Preserves current position and play/pause state. Concurrent calls are serialized. */
  async setStrategy(strategy, reason) {
    if (!this.mediaContext) throw new chunk2IJ66NTD_cjs.AvbridgeError(chunk2IJ66NTD_cjs.ERR_PLAYER_NOT_READY, "Player not ready \u2014 wait for the 'ready' event before calling playback methods.", "Await the 'ready' event or check player.readyState before calling play/pause/seek.");
    if (this.session?.strategy === strategy) return;
    this.switchingPromise = this.switchingPromise.then(
      () => this.doSetStrategy(strategy, reason)
    );
    await this.switchingPromise;
  }
  async doSetStrategy(strategy, reason) {
    const currentTime = this.session?.getCurrentTime() ?? 0;
    const wasPlaying = this.session ? !this.options.target.paused : false;
    const fromStrategy = this.session?.strategy ?? "native";
    const switchReason = reason ?? `manual switch to ${strategy}`;
    this.emitter.emit("strategychange", {
      from: fromStrategy,
      to: strategy,
      reason: switchReason,
      currentTime
    });
    this.diag.recordStrategySwitch(strategy, switchReason);
    this.clearSupervisor();
    if (this.session) {
      try {
        await this.session.destroy();
      } catch {
      }
      this.session = null;
    }
    const plugin = this.registry.findFor(this.mediaContext, strategy);
    if (!plugin) throw new Error(`no plugin available for strategy "${strategy}"`);
    this.session = await plugin.execute(this.mediaContext, this.options.target, this.transport);
    this.emitter.emitSticky("strategy", {
      strategy,
      reason: switchReason
    });
    this.session.onFatalError?.((fatalReason) => {
      void this.escalate(fatalReason);
    });
    this.attachSupervisor();
    try {
      await this.session.seek(currentTime);
      if (wasPlaying) await this.session.play();
    } catch (err) {
      console.warn("[avbridge] failed to restore position after strategy switch:", err);
    }
  }
  // ── Timeupdate loop ───────────────────────────────────────────────────
  startTimeupdateLoop() {
    this.timeupdateInterval = setInterval(() => {
      const t = this.session?.getCurrentTime() ?? this.options.target.currentTime;
      this.emitter.emit("timeupdate", { currentTime: t });
    }, 250);
  }
  // ── Public API ────────────────────────────────────────────────────────
  /** Subscribe to a player event. Returns an unsubscribe function. Sticky events (strategy, ready, tracks) replay for late subscribers. */
  on(event, fn) {
    return this.emitter.on(event, fn);
  }
  /** Remove a previously registered event listener. */
  off(event, fn) {
    this.emitter.off(event, fn);
  }
  /** Begin or resume playback. Throws if the player is not ready. */
  async play() {
    if (!this.session) throw new chunk2IJ66NTD_cjs.AvbridgeError(chunk2IJ66NTD_cjs.ERR_PLAYER_NOT_READY, "Player not ready \u2014 wait for the 'ready' event before calling playback methods.", "Await the 'ready' event or check player.readyState before calling play/pause/seek.");
    this.userIntent = "play";
    this.autoPausedForVisibility = false;
    await this.session.play();
  }
  /** Pause playback. No-op if the player is not ready or already paused. */
  pause() {
    this.userIntent = "pause";
    this.autoPausedForVisibility = false;
    this.session?.pause();
  }
  /**
   * Handle browser tab visibility changes. On hide: pause if the user
   * had been playing. On show: resume if we were the one who paused.
   * Skips when `backgroundBehavior: "continue"` is set (listener isn't
   * installed in that case).
   */
  onVisibilityChange() {
    if (!this.session) return;
    const action = decideVisibilityAction({
      hidden: document.hidden,
      userIntent: this.userIntent,
      sessionIsPlaying: !this.options.target.paused,
      autoPausedForVisibility: this.autoPausedForVisibility
    });
    if (action === "pause") {
      this.autoPausedForVisibility = true;
      chunkG4APZMCP_cjs.dbg.info("visibility", "tab hidden \u2014 auto-paused");
      this.session.pause();
    } else if (action === "resume") {
      this.autoPausedForVisibility = false;
      chunkG4APZMCP_cjs.dbg.info("visibility", "tab visible \u2014 auto-resuming");
      void this.session.play().catch((err) => {
        console.warn("[avbridge] auto-resume after tab return failed:", err);
      });
    }
  }
  /** Seek to the given time in seconds. Throws if the player is not ready. */
  async seek(time) {
    if (!this.session) throw new chunk2IJ66NTD_cjs.AvbridgeError(chunk2IJ66NTD_cjs.ERR_PLAYER_NOT_READY, "Player not ready \u2014 wait for the 'ready' event before calling playback methods.", "Await the 'ready' event or check player.readyState before calling play/pause/seek.");
    await this.session.seek(time);
  }
  /** Switch the active audio track by track ID. Throws if the player is not ready. */
  async setAudioTrack(id) {
    if (!this.session) throw new chunk2IJ66NTD_cjs.AvbridgeError(chunk2IJ66NTD_cjs.ERR_PLAYER_NOT_READY, "Player not ready \u2014 wait for the 'ready' event before calling playback methods.", "Await the 'ready' event or check player.readyState before calling play/pause/seek.");
    await this.session.setAudioTrack(id);
  }
  /** Switch the active subtitle track by track ID, or pass `null` to disable subtitles. */
  async setSubtitleTrack(id) {
    if (!this.session) throw new chunk2IJ66NTD_cjs.AvbridgeError(chunk2IJ66NTD_cjs.ERR_PLAYER_NOT_READY, "Player not ready \u2014 wait for the 'ready' event before calling playback methods.", "Await the 'ready' event or check player.readyState before calling play/pause/seek.");
    await this.session.setSubtitleTrack(id);
  }
  /** Return a snapshot of current diagnostics: container, codecs, strategy, runtime stats, and strategy history. */
  getDiagnostics() {
    if (this.session) {
      this.diag.recordRuntime(this.session.getRuntimeStats());
    }
    return this.diag.snapshot();
  }
  /** Return the total duration in seconds, or `NaN` if unknown. */
  getDuration() {
    const fromDiag = this.diag.snapshot().duration;
    if (typeof fromDiag === "number" && Number.isFinite(fromDiag)) return fromDiag;
    const fromVideo = this.options.target.duration;
    return Number.isFinite(fromVideo) ? fromVideo : NaN;
  }
  /** Return the current playback position in seconds. */
  getCurrentTime() {
    return this.session?.getCurrentTime() ?? this.options.target.currentTime ?? 0;
  }
  /** Tear down the player: stop timers, destroy the active session, remove all event listeners. The player is unusable after this call. */
  async destroy() {
    if (this.timeupdateInterval) {
      clearInterval(this.timeupdateInterval);
      this.timeupdateInterval = null;
    }
    this.clearSupervisor();
    if (this.endedListener) {
      this.options.target.removeEventListener("ended", this.endedListener);
      this.endedListener = null;
    }
    if (this.visibilityListener) {
      document.removeEventListener("visibilitychange", this.visibilityListener);
      this.visibilityListener = null;
    }
    if (this.session) {
      await this.session.destroy();
      this.session = null;
    }
    this.subtitleResources.revokeAll();
    this.emitter.removeAll();
  }
};
async function createPlayer(options) {
  return UnifiedPlayer.create(options);
}
function decideVisibilityAction(state) {
  if (state.hidden) {
    if (state.userIntent === "play" && state.sessionIsPlaying) return "pause";
    return "noop";
  }
  if (state.autoPausedForVisibility) return "resume";
  return "noop";
}
function buildInitialDecision(initial, ctx) {
  const natural = classifyContext(ctx);
  const cls = strategyToClass(initial, natural);
  const inherited = natural.fallbackChain ?? defaultFallbackChain(initial);
  const fallbackChain = inherited.filter((s) => s !== initial);
  return {
    class: cls,
    strategy: initial,
    reason: `initial strategy "${initial}" requested via options.initialStrategy`,
    fallbackChain
  };
}
function strategyToClass(strategy, natural) {
  if (natural.strategy === strategy) return natural.class;
  switch (strategy) {
    case "native":
      return "NATIVE";
    case "remux":
      return "REMUX_CANDIDATE";
    case "hybrid":
      return "HYBRID_CANDIDATE";
    case "fallback":
      return "FALLBACK_REQUIRED";
  }
}
function defaultFallbackChain(strategy) {
  switch (strategy) {
    case "native":
      return ["remux", "hybrid", "fallback"];
    case "remux":
      return ["hybrid", "fallback"];
    case "hybrid":
      return ["fallback"];
    case "fallback":
      return [];
  }
}

exports.FALLBACK_AUDIO_CODECS = FALLBACK_AUDIO_CODECS;
exports.FALLBACK_VIDEO_CODECS = FALLBACK_VIDEO_CODECS;
exports.NATIVE_AUDIO_CODECS = NATIVE_AUDIO_CODECS;
exports.NATIVE_VIDEO_CODECS = NATIVE_VIDEO_CODECS;
exports.UnifiedPlayer = UnifiedPlayer;
exports.classifyContext = classifyContext;
exports.createPlayer = createPlayer;
