// avbridge/dist/element.cjs
'use strict';

var chunkOFJYEITB_cjs = require('./chunk-OFJYEITB.cjs');
require('./chunk-WRKO6Q42.cjs');
require('./chunk-VLI3Y6IJ.cjs');
require('./chunk-2IJ66NTD.cjs');
require('./chunk-QDJLQR53.cjs');
require('./chunk-HZUVMXBN.cjs');
require('./chunk-YPZFGJV3.cjs');
require('./chunk-G4APZMCP.cjs');
require('./chunk-F3LQJKXK.cjs');

// src/element/avbridge-video.ts
var PREFERRED_STRATEGY_VALUES = /* @__PURE__ */ new Set([
  "auto",
  "native",
  "remux",
  "hybrid",
  "fallback"
]);
var FIT_VALUES = /* @__PURE__ */ new Set(["contain", "cover", "fill"]);
var DEFAULT_FIT = "contain";
var FORWARDED_VIDEO_EVENTS = [
  "loadstart",
  "loadedmetadata",
  "loadeddata",
  "canplay",
  "canplaythrough",
  "play",
  "playing",
  "pause",
  "seeking",
  "seeked",
  "volumechange",
  "ratechange",
  "durationchange",
  "waiting",
  "stalled",
  "emptied",
  "resize",
  "error"
];
var HTMLElementCtor = typeof HTMLElement !== "undefined" ? HTMLElement : class {
};
var AvbridgeVideoElement = class extends HTMLElementCtor {
  static observedAttributes = [
    "src",
    "autoplay",
    "muted",
    "loop",
    "preload",
    "poster",
    "playsinline",
    "crossorigin",
    "disableremoteplayback",
    "diagnostics",
    "preferstrategy",
    "fit",
    "no-orientation-lock"
  ];
  // ── Internal state ─────────────────────────────────────────────────────
  /** The shadow DOM `<video>` element that strategies render into. */
  _videoEl;
  /** Active player session, if any. Cleared on teardown. */
  _player = null;
  /**
   * Monotonic counter incremented on every (re)bootstrap. Async bootstrap
   * work captures the current ID; if it doesn't match by the time the work
   * resolves, the work is discarded.
   */
  _bootstrapId = 0;
  /** True after destroy() — element is permanently unusable. */
  _destroyed = false;
  /** Internal source state. Either string-form (src) OR rich (source). */
  _src = null;
  _source = null;
  /**
   * Set when the `source` property setter is in the middle of clearing the
   * `src` attribute as part of mutual exclusion. The attributeChangedCallback
   * checks this flag and skips its normal "clear source" side effect, which
   * would otherwise wipe the value we just set.
   */
  _suppressSrcAttrCallback = false;
  /** Last-known runtime state surfaced via getters. */
  _strategy = null;
  _strategyClass = null;
  _audioTracks = [];
  /** Subtitle tracks reported by the active UnifiedPlayer (options.subtitles
   *  + embedded container tracks + programmatic addSubtitle calls). */
  _subtitleTracks = [];
  /** Subtitle tracks derived from light-DOM `<track>` children. Maintained
   *  by _syncTextTracks on every mutation. Merged into the public
   *  `subtitleTracks` getter so the player's settings menu sees them. */
  _htmlTrackInfo = [];
  /**
   * External subtitle list forwarded to `createPlayer()` on the next
   * bootstrap. Setting this after bootstrap queues it for the next
   * source change; consumers that need to swap subtitles mid-playback
   * should set `source` to reload.
   */
  _subtitles = null;
  /**
   * Initial strategy preference. `"auto"` means "let the classifier decide";
   * any other value is passed to `createPlayer({ initialStrategy })` and
   * skips classification on the next bootstrap. Note that this only affects
   * the *initial* pick — runtime fallback escalation still applies, so a
   * preference of `"native"` may still escalate to remux/hybrid/fallback if
   * native fails.
   */
  _preferredStrategy = "auto";
  /** Current fit mode. Applied to the inner `<video>` via object-fit, and
   *  to the fallback canvas via the `--avbridge-fit` CSS custom property on
   *  the stage wrapper (see `src/strategies/fallback/video-renderer.ts`). */
  _fit = DEFAULT_FIT;
  /** The stage wrapper — the element the canvas attaches into, and where
   *  the `--avbridge-fit` CSS custom property lives. */
  _stageEl;
  /** Set if currentTime was assigned before the player was ready. */
  _pendingSeek = null;
  /** Set if play() was called before the player was ready. */
  _pendingPlay = false;
  /** MutationObserver tracking light-DOM `<track>` children. */
  _trackObserver = null;
  /** Document-level fullscreenchange handler — installed while connected so
   *  the element can lock/unlock screen orientation to match the video's
   *  intrinsic aspect. */
  _fullscreenChangeHandler = null;
  /** True if we successfully called screen.orientation.lock() on the last
   *  fullscreen entry. Used to know whether to unlock on exit. */
  _orientationLocked = false;
  // ── Construction & lifecycle ───────────────────────────────────────────
  constructor() {
    super();
    const root = this.attachShadow({ mode: "open" });
    const stage = document.createElement("div");
    stage.setAttribute("part", "stage");
    stage.style.cssText = `position:relative;width:100%;height:100%;display:block;--avbridge-fit:${DEFAULT_FIT};`;
    root.appendChild(stage);
    this._stageEl = stage;
    this._videoEl = document.createElement("video");
    this._videoEl.setAttribute("part", "video");
    this._videoEl.style.cssText = `width:100%;height:100%;display:block;background:#000;object-fit:var(--avbridge-fit, ${DEFAULT_FIT});`;
    this._videoEl.playsInline = true;
    stage.appendChild(this._videoEl);
    this._videoEl.addEventListener("progress", () => {
      if (this._destroyed) return;
      this._dispatch("progress", { buffered: this._videoEl.buffered });
    });
    for (const eventName of FORWARDED_VIDEO_EVENTS) {
      this._videoEl.addEventListener(eventName, () => {
        if (this._destroyed) return;
        this.dispatchEvent(new Event(eventName, { bubbles: false }));
      });
    }
  }
  connectedCallback() {
    if (this._destroyed) return;
    this._syncTextTracks();
    if (!this._trackObserver) {
      this._trackObserver = new MutationObserver(() => this._syncTextTracks());
      this._trackObserver.observe(this, { childList: true, subtree: false });
    }
    if (!this._fullscreenChangeHandler) {
      this._fullscreenChangeHandler = () => this._onFullscreenChange();
      document.addEventListener("fullscreenchange", this._fullscreenChangeHandler);
    }
    const source = this._activeSource();
    if (source != null) {
      void this._bootstrap(source);
    }
  }
  disconnectedCallback() {
    if (this._destroyed) return;
    if (this._trackObserver) {
      this._trackObserver.disconnect();
      this._trackObserver = null;
    }
    if (this._fullscreenChangeHandler) {
      document.removeEventListener("fullscreenchange", this._fullscreenChangeHandler);
      this._fullscreenChangeHandler = null;
    }
    this._releaseOrientationLock();
    this._bootstrapId++;
    void this._teardown();
  }
  attributeChangedCallback(name, _oldValue, newValue) {
    if (this._destroyed) return;
    switch (name) {
      case "src":
        if (this._suppressSrcAttrCallback) break;
        this._setSrcInternal(newValue);
        break;
      case "autoplay":
      case "muted":
      case "loop":
      case "playsinline":
      case "disableremoteplayback":
        if (newValue == null) this._videoEl.removeAttribute(name);
        else this._videoEl.setAttribute(name, newValue);
        break;
      case "preload":
      case "poster":
      case "crossorigin":
        if (newValue == null) this._videoEl.removeAttribute(name);
        else this._videoEl.setAttribute(name, newValue);
        break;
      case "diagnostics":
        break;
      case "preferstrategy":
        if (newValue && PREFERRED_STRATEGY_VALUES.has(newValue)) {
          this._preferredStrategy = newValue;
        } else {
          this._preferredStrategy = "auto";
        }
        break;
      case "fit": {
        const next = newValue && FIT_VALUES.has(newValue) ? newValue : DEFAULT_FIT;
        if (next === this._fit) break;
        this._fit = next;
        this._stageEl.style.setProperty("--avbridge-fit", next);
        this._dispatch("fitchange", { fit: next });
        break;
      }
    }
  }
  // ── Source handling ────────────────────────────────────────────────────
  /** Returns the currently-active source (src or source), whichever is set. */
  _activeSource() {
    if (this._source != null) return this._source;
    if (this._src != null) return this._src;
    return null;
  }
  /**
   * Mirror light-DOM `<track>` children into the shadow `<video>` so that
   * the browser's native text-track machinery picks them up. Called on
   * connect, on every mutation of light-DOM children, and once after each
   * source change so newly-set tracks survive a fresh `<video>`.
   *
   * Strategy: clone the children. We don't move them because the user's
   * code may still hold references to the originals (e.g. to set `default`).
   * The shadow copies are throwaway — we wipe them on every sync.
   */
  _syncTextTracks() {
    const existing = this._videoEl.querySelectorAll("track");
    for (const t of Array.from(existing)) t.remove();
    this._htmlTrackInfo = [];
    let htmlIdx = 0;
    for (const child of Array.from(this.children)) {
      if (child.tagName === "TRACK") {
        const track = child;
        const clone = track.cloneNode(true);
        this._videoEl.appendChild(clone);
        const src = track.getAttribute("src") ?? void 0;
        const format = src?.toLowerCase().endsWith(".srt") ? "srt" : "vtt";
        this._htmlTrackInfo.push({
          id: 1e4 + htmlIdx,
          format,
          language: track.srclang || track.getAttribute("label") || void 0,
          sidecarUrl: src
        });
        htmlIdx++;
      }
    }
    this._dispatch("trackschange", {
      audioTracks: this._audioTracks,
      subtitleTracks: this.subtitleTracks
    });
  }
  /** Internal src setter — separate from the property setter so the
   * attributeChangedCallback can use it without re-entering reflection. */
  _setSrcInternal(value) {
    if (value === this._src && this._source == null) return;
    this._src = value;
    this._source = null;
    this._onSourceChanged();
  }
  /** Called whenever the active source changes (src or source). */
  _onSourceChanged() {
    if (this._destroyed) return;
    const source = this._activeSource();
    if (source == null) {
      this._bootstrapId++;
      void this._teardown();
      return;
    }
    if (this.isConnected) {
      void this._bootstrap(source);
    }
  }
  // ── Bootstrap (the only place a UnifiedPlayer is created) ──────────────
  async _bootstrap(source) {
    if (this._destroyed) return;
    const id = ++this._bootstrapId;
    await this._teardown(id);
    if (id !== this._bootstrapId || this._destroyed) return;
    this._dispatch("loadstart", {});
    let player;
    try {
      player = await chunkOFJYEITB_cjs.createPlayer({
        source,
        target: this._videoEl,
        // Honor the consumer's preferred initial strategy. "auto" means
        // "let the classifier decide" — the createPlayer call simply doesn't
        // pass initialStrategy in that case.
        ...this._preferredStrategy !== "auto" ? { initialStrategy: this._preferredStrategy } : {},
        ...this._subtitles ? { subtitles: this._subtitles } : {}
      });
    } catch (err) {
      if (id !== this._bootstrapId || this._destroyed) return;
      this._dispatchError(err);
      return;
    }
    if (id !== this._bootstrapId || this._destroyed || !this.isConnected) {
      try {
        await player.destroy();
      } catch {
      }
      return;
    }
    this._player = player;
    this._syncTextTracks();
    player.on("strategy", ({ strategy, reason }) => {
      const cls = player.getDiagnostics().strategyClass;
      this._strategy = strategy;
      this._strategyClass = cls === "pending" ? null : cls;
      this._dispatch("strategychange", {
        strategy,
        strategyClass: this._strategyClass,
        reason,
        diagnostics: player.getDiagnostics()
      });
    });
    player.on("strategychange", ({ from, to, reason, currentTime }) => {
      this._dispatch("strategychange", {
        from,
        strategy: to,
        strategyClass: player.getDiagnostics().strategyClass === "pending" ? null : player.getDiagnostics().strategyClass,
        reason,
        currentTime,
        diagnostics: player.getDiagnostics()
      });
    });
    player.on("tracks", ({ video: _v, audio, subtitle }) => {
      this._audioTracks = audio;
      this._subtitleTracks = subtitle;
      this._dispatch("trackschange", {
        audioTracks: audio,
        subtitleTracks: subtitle
      });
    });
    player.on("error", (err) => {
      this._dispatchError(err);
    });
    player.on("timeupdate", ({ currentTime }) => {
      this._dispatch("timeupdate", { currentTime });
    });
    player.on("ended", () => {
      this._dispatch("ended", {});
    });
    player.on("ready", () => {
      this._dispatch("ready", { diagnostics: player.getDiagnostics() });
      if (this._pendingSeek != null) {
        const t = this._pendingSeek;
        this._pendingSeek = null;
        void player.seek(t).catch(() => {
        });
      }
      if (this._pendingPlay) {
        this._pendingPlay = false;
        void player.play().catch(() => {
        });
      } else if (this.autoplay) {
        void player.play().catch(() => {
        });
      }
    });
  }
  /**
   * Tear down the active player and reset runtime state. Idempotent.
   * If `currentBootstrapId` is provided, the bootstrap counter is NOT
   * incremented (used by `_bootstrap()` to avoid invalidating itself).
   */
  async _teardown(currentBootstrapId) {
    if (currentBootstrapId == null) {
      this._bootstrapId++;
    }
    const player = this._player;
    this._player = null;
    this._strategy = null;
    this._strategyClass = null;
    this._audioTracks = [];
    this._subtitleTracks = [];
    if (player) {
      try {
        await player.destroy();
      } catch {
      }
    }
  }
  // ── Public properties ──────────────────────────────────────────────────
  get src() {
    return this._src;
  }
  set src(value) {
    if (value == null) {
      this.removeAttribute("src");
    } else {
      this.setAttribute("src", value);
    }
  }
  get source() {
    return this._source;
  }
  set source(value) {
    if (value === this._source && this._src == null) return;
    this._source = value;
    if (value != null) {
      this._src = null;
      if (this.hasAttribute("src")) {
        this._suppressSrcAttrCallback = true;
        try {
          this.removeAttribute("src");
        } finally {
          this._suppressSrcAttrCallback = false;
        }
      }
    }
    this._onSourceChanged();
  }
  get autoplay() {
    return this.hasAttribute("autoplay");
  }
  set autoplay(value) {
    if (value) this.setAttribute("autoplay", "");
    else this.removeAttribute("autoplay");
  }
  get muted() {
    return this._videoEl.muted;
  }
  set muted(value) {
    this._videoEl.muted = value;
    if (value) this.setAttribute("muted", "");
    else this.removeAttribute("muted");
  }
  get loop() {
    return this.hasAttribute("loop");
  }
  set loop(value) {
    if (value) this.setAttribute("loop", "");
    else this.removeAttribute("loop");
  }
  get preload() {
    const v = this.getAttribute("preload");
    return v === "none" || v === "metadata" || v === "auto" ? v : "auto";
  }
  set preload(value) {
    this.setAttribute("preload", value);
  }
  get diagnostics() {
    return this.hasAttribute("diagnostics");
  }
  set diagnostics(value) {
    if (value) this.setAttribute("diagnostics", "");
    else this.removeAttribute("diagnostics");
  }
  get fit() {
    return this._fit;
  }
  set fit(value) {
    if (!FIT_VALUES.has(value)) return;
    this.setAttribute("fit", value);
  }
  get preferredStrategy() {
    return this._preferredStrategy;
  }
  set preferredStrategy(value) {
    if (PREFERRED_STRATEGY_VALUES.has(value)) {
      this.setAttribute("preferstrategy", value);
    }
  }
  get currentTime() {
    return this._player?.getCurrentTime() ?? 0;
  }
  set currentTime(value) {
    if (this._player) {
      void this._player.seek(value).catch(() => {
      });
    } else {
      this._pendingSeek = value;
    }
  }
  get duration() {
    return this._player?.getDuration() ?? NaN;
  }
  get paused() {
    return this._videoEl.paused;
  }
  get ended() {
    return this._videoEl.ended;
  }
  get readyState() {
    return this._videoEl.readyState;
  }
  /**
   * Buffered time ranges for the active source. Mirrors the standard
   * `<video>.buffered` `TimeRanges` API.
   *
   * - **Native / remux:** pass-through to the real `<video>.buffered`
   *   (reflects the browser's SourceBuffer / progressive-download state).
   * - **Hybrid / fallback:** a single `[0, frontier]` range synthesized
   *   from the demuxer's read progress — "how far libav has ever pumped
   *   packets through." Monotonic; does not shrink on seek. This is an
   *   approximation, not MSE-fidelity: decoded frames on canvas strategies
   *   are consumed in flight, so we can't report per-range availability
   *   the way MSE does. Enough for a seek-bar buffered indicator.
   */
  get buffered() {
    return this._videoEl.buffered;
  }
  // ── HTMLMediaElement parity ───────────────────────────────────────────
  // Mirror the standard <video> surface so consumers can drop the element
  // in as a <video> replacement. Each property is a thin passthrough to the
  // shadow `<video>`.
  get poster() {
    return this._videoEl.poster;
  }
  set poster(value) {
    if (value == null || value === "") this.removeAttribute("poster");
    else this.setAttribute("poster", value);
  }
  get volume() {
    return this._videoEl.volume;
  }
  set volume(value) {
    this._videoEl.volume = value;
  }
  get playbackRate() {
    return this._videoEl.playbackRate;
  }
  set playbackRate(value) {
    this._videoEl.playbackRate = value;
  }
  get videoWidth() {
    return this._videoEl.videoWidth;
  }
  get videoHeight() {
    return this._videoEl.videoHeight;
  }
  get played() {
    return this._videoEl.played;
  }
  get seekable() {
    return this._videoEl.seekable;
  }
  get crossOrigin() {
    return this._videoEl.crossOrigin;
  }
  set crossOrigin(value) {
    if (value == null) this.removeAttribute("crossorigin");
    else this.setAttribute("crossorigin", value);
  }
  get disableRemotePlayback() {
    return this._videoEl.disableRemotePlayback;
  }
  set disableRemotePlayback(value) {
    if (value) this.setAttribute("disableremoteplayback", "");
    else this.removeAttribute("disableremoteplayback");
  }
  /**
   * Native `HTMLMediaElement.canPlayType()` passthrough. Note that this
   * answers about the *browser's* native support, not avbridge's full
   * capabilities — avbridge can play many formats this method returns ""
   * for, by routing them to the remux/hybrid/fallback strategies.
   */
  canPlayType(mimeType) {
    return this._videoEl.canPlayType(mimeType);
  }
  /**
   * **Escape hatch.** The underlying shadow-DOM `<video>` element.
   *
   * Use for native browser APIs the wrapper doesn't expose:
   * - `el.videoElement.requestPictureInPicture()`
   * - `el.videoElement.audioTracks` (browser native, not avbridge's track list)
   * - direct integration with libraries that need a real HTMLVideoElement
   *
   * **Caveat:** When the active strategy is `"fallback"` or `"hybrid"`,
   * frames are rendered to a canvas overlay, not into this `<video>`.
   * APIs that depend on the actual pixels (Picture-in-Picture, captureStream)
   * will not show the playing content in those modes. Check `el.strategy`
   * before using such APIs.
   */
  get videoElement() {
    return this._videoEl;
  }
  get strategy() {
    return this._strategy;
  }
  get strategyClass() {
    return this._strategyClass;
  }
  get player() {
    return this._player;
  }
  get audioTracks() {
    return this._audioTracks;
  }
  get subtitleTracks() {
    return this._htmlTrackInfo.length === 0 ? this._subtitleTracks : [...this._subtitleTracks, ...this._htmlTrackInfo];
  }
  /**
   * External subtitle files to attach when the source loads. Takes effect
   * on the next bootstrap — set before assigning `source`, or reload via
   * `load()` after changing. For dynamic post-bootstrap addition, use
   * `addSubtitle()` instead.
   *
   * @example
   * el.subtitles = [{ url: "/en.srt", format: "srt", language: "en" }];
   * el.src = "/movie.mp4";
   */
  get subtitles() {
    return this._subtitles;
  }
  set subtitles(value) {
    this._subtitles = value;
  }
  /**
   * Attach a subtitle track to the current playback without rebuilding
   * the player. Works while the element is playing — converts SRT to
   * VTT if needed, adds a `<track>` to the inner `<video>`. Canvas
   * strategies pick up the new track via their textTracks watcher.
   */
  async addSubtitle(subtitle) {
    const { attachSubtitleTracks } = await import('./subtitles-HMVGWTU2.cjs');
    const format = subtitle.format ?? (subtitle.url.endsWith(".srt") ? "srt" : "vtt");
    const track = {
      id: this._subtitleTracks.length,
      format,
      language: subtitle.language,
      sidecarUrl: subtitle.url
    };
    this._subtitleTracks.push(track);
    console.log(`[avbridge:subs] addSubtitle id=${track.id} format=${format} lang=${subtitle.language ?? "?"}`);
    await attachSubtitleTracks(
      this._videoEl,
      this._subtitleTracks,
      void 0,
      (err, t) => {
        console.warn(`[avbridge:subs] subtitle ${t.id} failed: ${err.message}`);
      }
    );
    const textTracks = this._videoEl.textTracks;
    for (let i = 0; i < textTracks.length; i++) {
      if (textTracks[i].label === (subtitle.language ?? `Subtitle ${track.id}`)) {
        textTracks[i].mode = "showing";
        console.log(`[avbridge:subs] enabled textTrack[${i}] mode=showing`);
        break;
      }
    }
    this._dispatch("trackschange", {
      audioTracks: this._audioTracks,
      subtitleTracks: this.subtitleTracks
    });
  }
  /**
   * Disable the automatic `screen.orientation.lock()` that runs on
   * fullscreen entry. Set when you want to honor the device's native
   * auto-rotate instead of matching the video's intrinsic orientation.
   */
  get noOrientationLock() {
    return this.hasAttribute("no-orientation-lock");
  }
  set noOrientationLock(value) {
    if (value) this.setAttribute("no-orientation-lock", "");
    else this.removeAttribute("no-orientation-lock");
  }
  // ── Fullscreen orientation lock ────────────────────────────────────────
  /** Called whenever `document.fullscreenchange` fires. If this element (or
   *  any of its ancestors) is now fullscreen, derive the target orientation
   *  from the video's intrinsic size and call `screen.orientation.lock()`.
   *  On exit, release the lock we took. iOS Safari rejects `lock()` — we
   *  swallow the rejection so nothing breaks on that path. */
  _onFullscreenChange() {
    if (this._destroyed) return;
    const fsEl = document.fullscreenElement;
    const nowFullscreen = fsEl != null && this._isInsideOrEquals(fsEl);
    if (nowFullscreen && !this._orientationLocked) {
      if (this.noOrientationLock) return;
      const target = this._desiredOrientation();
      if (!target) return;
      void this._lockOrientation(target);
    } else if (!nowFullscreen && this._orientationLocked) {
      this._releaseOrientationLock();
    }
  }
  /** Walk composed-tree ancestors to see if `target` is this element or
   *  any ancestor across shadow boundaries. `Node.contains()` can't cross
   *  shadow roots, so when `<avbridge-player>` (the fullscreen element)
   *  hosts this `<avbridge-video>` inside its shadow DOM, `contains()`
   *  returns false. */
  _isInsideOrEquals(target) {
    let node = this;
    while (node) {
      if (node === target) return true;
      const parent = node.parentNode;
      if (parent instanceof ShadowRoot) node = parent.host;
      else node = parent;
    }
    return false;
  }
  /** Derive "landscape" / "portrait" from the intrinsic video dimensions.
   *  Returns null when dimensions aren't known yet or the video is square.
   *  Uses `videoWidth` / `videoHeight` from the inner `<video>`, which the
   *  browser sets to the display-aspect-corrected size (so anamorphic
   *  content is judged by its display aspect, not pixel aspect). */
  _desiredOrientation() {
    const w = this._videoEl.videoWidth;
    const h = this._videoEl.videoHeight;
    if (!w || !h) return null;
    if (w === h) return null;
    return w > h ? "landscape" : "portrait";
  }
  /** Attempt to lock screen orientation. Swallows rejections — iOS Safari
   *  doesn't implement `lock()`, and desktop / non-fullscreen contexts will
   *  reject too. Records success so we know whether to unlock on exit. */
  async _lockOrientation(target) {
    const so = screen.orientation;
    if (!so || typeof so.lock !== "function") return;
    try {
      await so.lock(target);
      this._orientationLocked = true;
    } catch {
    }
  }
  _releaseOrientationLock() {
    if (!this._orientationLocked) return;
    this._orientationLocked = false;
    const so = screen.orientation;
    if (so && typeof so.unlock === "function") {
      try {
        so.unlock();
      } catch {
      }
    }
  }
  // ── Public methods ─────────────────────────────────────────────────────
  /** Force a (re-)bootstrap if a source is currently set. */
  async load() {
    if (this._destroyed) return;
    const source = this._activeSource();
    if (source == null) return;
    await this._bootstrap(source);
  }
  /**
   * Begin or resume playback. If the player isn't ready yet, the call is
   * queued and applied once `ready` fires.
   */
  async play() {
    if (this._destroyed) return;
    if (this._player) {
      await this._player.play();
    } else {
      this._pendingPlay = true;
    }
  }
  pause() {
    if (this._destroyed) return;
    this._pendingPlay = false;
    this._player?.pause();
  }
  /**
   * Tear down the element permanently. After destroy(), the element ignores
   * all method calls and attribute changes.
   */
  async destroy() {
    if (this._destroyed) return;
    this._destroyed = true;
    await this._teardown();
    this._dispatch("destroy", {});
  }
  async setAudioTrack(id) {
    if (this._destroyed || !this._player) return;
    await this._player.setAudioTrack(id);
  }
  async setSubtitleTrack(id) {
    if (this._destroyed || !this._player) return;
    await this._player.setSubtitleTrack(id);
  }
  getDiagnostics() {
    return this._player?.getDiagnostics() ?? null;
  }
  addEventListener(type, listener, options) {
    super.addEventListener(type, listener, options);
  }
  removeEventListener(type, listener, options) {
    super.removeEventListener(type, listener, options);
  }
  // ── Event helpers ──────────────────────────────────────────────────────
  _dispatch(name, detail) {
    this.dispatchEvent(new CustomEvent(name, { detail, bubbles: false }));
  }
  _dispatchError(err) {
    const error = err instanceof Error ? err : new Error(String(err));
    this._dispatch("error", { error, diagnostics: this._player?.getDiagnostics() ?? null });
  }
};

// src/element.ts
if (typeof customElements !== "undefined" && !customElements.get("avbridge-video")) {
  customElements.define("avbridge-video", AvbridgeVideoElement);
}

exports.AvbridgeVideoElement = AvbridgeVideoElement;
