// avbridge/dist/index.js
import { mimeForFormat, generateFilename, createOutputFormat } from './chunk-B76QWPFM.js';
export { remux } from './chunk-B76QWPFM.js';
export { FALLBACK_AUDIO_CODECS, FALLBACK_VIDEO_CODECS, NATIVE_AUDIO_CODECS, NATIVE_VIDEO_CODECS, UnifiedPlayer, classifyContext as classify, createPlayer } from './chunk-VOC24LYF.js';
export { srtToVtt } from './chunk-EDDWAN2L.js';
import { probe, buildMediabunnySourceFromInput } from './chunk-5CX7BVVV.js';
export { probe } from './chunk-5CX7BVVV.js';
import { AvbridgeError, ERR_CONTAINER_NOT_SUPPORTED, ERR_TRANSCODE_UNSUPPORTED_COMBO, ERR_TRANSCODE_ABORTED, ERR_TRANSCODE_DECODE, ERR_CODEC_NOT_SUPPORTED } from './chunk-CPJLFFCC.js';
export { AvbridgeError, ERR_ALL_STRATEGIES_EXHAUSTED, ERR_CODEC_NOT_SUPPORTED, ERR_FETCH_FAILED, ERR_LIBAV_NOT_REACHABLE, ERR_MSE_CODEC_NOT_SUPPORTED, ERR_MSE_NOT_SUPPORTED, ERR_PLAYER_NOT_READY, ERR_PROBE_FAILED, ERR_PROBE_FETCH_FAILED, ERR_PROBE_UNKNOWN_CONTAINER, ERR_RANGE_NOT_SUPPORTED, ERR_STRATEGY_FAILED } from './chunk-CPJLFFCC.js';
import './chunk-LUFA47FP.js';
import './chunk-3YKWU4FM.js';
import './chunk-3AI5WFFN.js';
import './chunk-5DMTJVIU.js';
import './chunk-5YAWWKA3.js';

// src/convert/transcode-libav.ts
function isLibavTranscodeContainer(container) {
  return container === "avi" || container === "asf" || container === "flv" || container === "rm";
}
async function transcodeViaLibav(ctx, options) {
  const outputFormat = options.outputFormat ?? "mp4";
  if (outputFormat !== "mp4" && outputFormat !== "webm" && outputFormat !== "mkv") {
    throw new AvbridgeError(
      ERR_TRANSCODE_UNSUPPORTED_COMBO,
      `legacy-container transcode supports MP4, WebM, and MKV output (got "${outputFormat}").`,
      `Use outputFormat: "mp4", "webm", or "mkv".`
    );
  }
  const videoCodec = options.videoCodec ?? (outputFormat === "webm" ? "vp9" : "h264");
  const audioCodec = options.audioCodec ?? (outputFormat === "webm" ? "opus" : "aac");
  const quality = options.quality ?? "medium";
  options.signal?.throwIfAborted();
  const [
    mb,
    { openLibavDemux, sanitizePacketTimestamp, sanitizeFrameTimestamp, libavFrameToInterleavedFloat32 },
    { normalizeSource },
    { createOutputFormat: createOutputFormat2, mimeForFormat: mimeForFormat2, generateFilename: generateFilename2 }
  ] = await Promise.all([
    import('mediabunny'),
    import('./libav-demux-SXZDLC7W.js'),
    import('./source-4TZ6KMNV.js'),
    import('./remux-PHUHO3VV.js')
  ]);
  const normalized = await normalizeSource(ctx.source);
  const demux = await openLibavDemux({
    source: normalized,
    filename: ctx.name ?? "input.bin",
    context: ctx
    // transport config is not yet threaded through ConvertOptions; add
    // later if URL-source transcode with signed URLs becomes a need.
  });
  try {
    let throwIfAborted2 = function() {
      if (ac?.aborted) {
        throw new AvbridgeError(
          ERR_TRANSCODE_ABORTED,
          "transcode: aborted by caller.",
          void 0
        );
      }
    }, throwIfDrainError2 = function() {
      if (drainError) {
        const msg = drainError.message;
        throw new AvbridgeError(
          ERR_TRANSCODE_DECODE,
          `transcode: video decoder error: ${msg}`,
          "This usually indicates the WebCodecs decoder rejected a malformed packet."
        );
      }
    };
    var throwIfAborted = throwIfAborted2, throwIfDrainError = throwIfDrainError2;
    options.signal?.throwIfAborted();
    if (!demux.videoStream && !demux.audioStream) {
      throw new Error("transcode: source has no decodable tracks");
    }
    if (options.outputStream) {
      throw new AvbridgeError(
        ERR_TRANSCODE_UNSUPPORTED_COMBO,
        "outputStream is not yet supported for the libav-backed transcode path.",
        "Remove the outputStream option to receive the transcoded blob in memory. Streaming output for this path is on the roadmap."
      );
    }
    const bufferTarget = new mb.BufferTarget();
    const output = new mb.Output({
      format: createOutputFormat2(mb, outputFormat),
      target: bufferTarget
    });
    const bridge = await loadBridge();
    let videoDecoder = null;
    let videoSoftDec = null;
    let videoSource = null;
    let videoBsfCtx = null;
    let videoBsfPkt = null;
    let videoWidth = 0;
    let videoHeight = 0;
    let videoTimeBase;
    const frameQueue = [];
    const MAX_QUEUE = 16;
    let draining = false;
    let drainError = null;
    let activeDrain = null;
    const drain = () => {
      if (draining) return activeDrain ?? Promise.resolve();
      draining = true;
      const run = (async () => {
        try {
          while (frameQueue.length > 0 && !drainError) {
            const frame = frameQueue.shift();
            try {
              const sample = new mb.VideoSample(frame, {
                timestamp: (frame.timestamp ?? 0) / 1e6
                // µs → s
              });
              await videoSource.add(sample);
            } finally {
              frame.close();
            }
          }
        } catch (err) {
          drainError = err;
          while (frameQueue.length > 0) {
            try {
              frameQueue.shift().close();
            } catch {
            }
          }
        } finally {
          draining = false;
          activeDrain = null;
        }
      })();
      activeDrain = run;
      return run;
    };
    if (demux.videoStream && !options.dropVideo) {
      try {
        const bitDepth = ctx.videoTracks[0]?.bitDepth ?? 8;
        if (bitDepth > 8) {
          throw new AvbridgeError(
            ERR_TRANSCODE_UNSUPPORTED_COMBO,
            `transcode: 10-bit video is not supported in this release (source bit depth: ${bitDepth}).`,
            `Phase 1 transcode handles 8-bit video only. 10-bit support is on the roadmap.`
          );
        }
        if (demux.videoStream.time_base_num && demux.videoStream.time_base_den) {
          videoTimeBase = [demux.videoStream.time_base_num, demux.videoStream.time_base_den];
        }
        let config = null;
        try {
          config = await bridge.videoStreamToConfig(demux.libav, demux.videoStream);
        } catch {
          config = null;
        }
        const supported = config ? await VideoDecoder.isConfigSupported(config).catch(() => ({ supported: false })) : { supported: false };
        videoWidth = config?.codedWidth ?? ctx.videoTracks[0]?.width ?? 0;
        videoHeight = config?.codedHeight ?? ctx.videoTracks[0]?.height ?? 0;
        if (config && supported.supported) {
          videoDecoder = new VideoDecoder({
            output: (frame) => {
              if (frameQueue.length >= MAX_QUEUE) {
                frame.close();
                return;
              }
              frameQueue.push(frame);
              void drain();
            },
            error: (err) => {
              drainError = err;
            }
          });
          videoDecoder.configure(config);
        } else {
          const libavSoft = demux.libav;
          const [, c, pkt, frame] = await libavSoft.ff_init_decoder(
            demux.videoStream.codec_id,
            { codecpar: demux.videoStream.codecpar }
          );
          videoSoftDec = { c, pkt, frame };
        }
        videoSource = new mb.VideoSampleSource({
          codec: avbridgeVideoToMediabunny(videoCodec),
          bitrate: qualityToMediabunny(mb, quality, options.videoBitrate),
          ...options.frameRate !== void 0 ? { frameRate: options.frameRate } : {},
          ...options.hardwareAcceleration !== void 0 ? { hardwareAcceleration: options.hardwareAcceleration } : {},
          // Progress reporting: media-time-based via each encoded packet.
          onEncodedPacket: options.onProgress ? (packet) => {
            const t = packet.timestamp;
            if (Number.isFinite(t) && ctx.duration && ctx.duration > 0) {
              const pct = Math.min(100, t / ctx.duration * 100);
              options.onProgress({ percent: pct, bytesWritten: 0 });
            }
          } : void 0
        });
        const videoMeta = {};
        if (options.width !== void 0) videoMeta.width = options.width;
        else if (videoWidth > 0) videoMeta.width = videoWidth;
        if (options.height !== void 0) videoMeta.height = options.height;
        else if (videoHeight > 0) videoMeta.height = videoHeight;
        if (options.frameRate !== void 0) videoMeta.frameRate = options.frameRate;
        output.addVideoTrack(videoSource, videoMeta);
        if (ctx.videoTracks[0]?.codec === "mpeg4") {
          const runtime = demux.libav;
          try {
            videoBsfCtx = await runtime.av_bsf_list_parse_str_js("mpeg4_unpack_bframes");
            if (videoBsfCtx != null && videoBsfCtx >= 0) {
              const parIn = await runtime.AVBSFContext_par_in(videoBsfCtx);
              await runtime.avcodec_parameters_copy(parIn, demux.videoStream.codecpar);
              await runtime.av_bsf_init(videoBsfCtx);
              videoBsfPkt = await demux.libav.av_packet_alloc();
            } else {
              videoBsfCtx = null;
            }
          } catch {
            videoBsfCtx = null;
          }
        }
      } catch (err) {
        if (err instanceof AvbridgeError) throw err;
        throw new AvbridgeError(
          ERR_CODEC_NOT_SUPPORTED,
          `transcode: video decoder init failed: ${err.message}`,
          `The source's video codec may not be supported by this browser's WebCodecs implementation.`
        );
      }
    }
    let audioDec = null;
    let audioSource = null;
    let audioTimeBase;
    const includeAudio = demux.audioStream && !options.dropAudio;
    if (includeAudio) {
      try {
        const libav = demux.libav;
        const [, c, pkt, frame] = await libav.ff_init_decoder(
          demux.audioStream.codec_id,
          { codecpar: demux.audioStream.codecpar }
        );
        audioDec = { c, pkt, frame };
        if (demux.audioStream.time_base_num && demux.audioStream.time_base_den) {
          audioTimeBase = [
            demux.audioStream.time_base_num,
            demux.audioStream.time_base_den
          ];
        }
        audioSource = new mb.AudioSampleSource({
          codec: avbridgeAudioToMediabunny(audioCodec),
          bitrate: qualityToMediabunny(mb, quality, options.audioBitrate)
        });
        output.addAudioTrack(audioSource);
      } catch (err) {
        const codecName = ctx.audioTracks[0]?.codec ?? "unknown";
        throw new AvbridgeError(
          ERR_CODEC_NOT_SUPPORTED,
          `transcode: no decoder available for audio codec "${codecName}" in this libav variant (${err.message}).`,
          `The file may still play via createPlayer() (fallback strategy). Pass { dropAudio: true } to transcode video-only.`
        );
      }
    } else if (options.dropAudio) {
    }
    if (!videoSource && !audioSource) {
      throw new AvbridgeError(
        ERR_TRANSCODE_UNSUPPORTED_COMBO,
        "transcode: no video or audio track to encode (did you set both dropVideo and dropAudio?).",
        "Remove dropVideo or dropAudio to include at least one track."
      );
    }
    await output.start();
    const videoFps = ctx.videoTracks[0]?.fps && ctx.videoTracks[0].fps > 0 ? ctx.videoTracks[0].fps : 30;
    const videoFrameStepUs = Math.max(1, Math.round(1e6 / videoFps));
    let syntheticVideoUs = 0;
    let syntheticAudioUs = 0;
    const libavFull = demux.libav;
    async function applyBSF(packets) {
      if (!videoBsfCtx || !videoBsfPkt) return packets;
      const out = [];
      for (const pkt of packets) {
        await libavFull.ff_copyin_packet(videoBsfPkt, pkt);
        const sendErr = await libavFull.av_bsf_send_packet(videoBsfCtx, videoBsfPkt);
        if (sendErr < 0) {
          out.push(pkt);
          continue;
        }
        while (true) {
          const recvErr = await libavFull.av_bsf_receive_packet(videoBsfCtx, videoBsfPkt);
          if (recvErr < 0) break;
          out.push(await libavFull.ff_copyout_packet(videoBsfPkt));
        }
      }
      return out;
    }
    const ac = options.signal;
    const onVideoPacketsWebCodecs = videoDecoder ? async (pkts) => {
      throwIfAborted2();
      throwIfDrainError2();
      while (!ac?.aborted && (videoDecoder.decodeQueueSize > 16 || frameQueue.length >= MAX_QUEUE - 2)) {
        await new Promise((r) => setTimeout(r, 10));
      }
      throwIfAborted2();
      const processed = await applyBSF(pkts);
      const bridgeAny = bridge;
      for (const pkt of processed) {
        sanitizePacketTimestamp(pkt, () => {
          const ts = syntheticVideoUs;
          syntheticVideoUs += videoFrameStepUs;
          return ts;
        }, videoTimeBase);
        try {
          const chunk = bridgeAny.packetToEncodedVideoChunk(pkt, demux.videoStream);
          videoDecoder.decode(chunk);
        } catch (err) {
          throw new AvbridgeError(
            ERR_TRANSCODE_DECODE,
            `transcode: packet \u2192 EncodedVideoChunk failed: ${err.message}`,
            void 0
          );
        }
      }
    } : void 0;
    const onVideoPacketsSoftware = videoSoftDec ? async (pkts) => {
      throwIfAborted2();
      throwIfDrainError2();
      while (!ac?.aborted && frameQueue.length >= MAX_QUEUE - 2) {
        await new Promise((r) => setTimeout(r, 10));
      }
      throwIfAborted2();
      const libavSoft = demux.libav;
      let frames;
      try {
        frames = await libavSoft.ff_decode_multi(
          videoSoftDec.c,
          videoSoftDec.pkt,
          videoSoftDec.frame,
          pkts,
          { ignoreErrors: true }
        );
      } catch (err) {
        throw new AvbridgeError(
          ERR_TRANSCODE_DECODE,
          `transcode: software video decode failed: ${err.message}`,
          void 0
        );
      }
      for (const f of frames) {
        sanitizeFrameTimestamp(f, () => {
          const ts = syntheticVideoUs;
          syntheticVideoUs += videoFrameStepUs;
          return ts;
        }, videoTimeBase);
        try {
          const vf = bridge.laFrameToVideoFrame(f, { timeBase: [1, 1e6] });
          if (frameQueue.length >= MAX_QUEUE) {
            vf.close();
          } else {
            frameQueue.push(vf);
            void drain();
          }
        } catch (err) {
          throw new AvbridgeError(
            ERR_TRANSCODE_DECODE,
            `transcode: laFrameToVideoFrame failed: ${err.message}`,
            void 0
          );
        }
      }
    } : void 0;
    await demux.pump({
      signal: ac,
      onVideoPackets: onVideoPacketsWebCodecs ?? onVideoPacketsSoftware,
      onAudioPackets: audioDec ? async (pkts) => {
        throwIfAborted2();
        await decodeAudioBatch(pkts, false);
      } : void 0,
      onEof: async () => {
        if (videoDecoder && videoDecoder.state === "configured") {
          try {
            await videoDecoder.flush();
          } catch {
          }
        }
        if (videoSoftDec) {
          const libavSoft = demux.libav;
          try {
            const tail = await libavSoft.ff_decode_multi(
              videoSoftDec.c,
              videoSoftDec.pkt,
              videoSoftDec.frame,
              [],
              { fin: true, ignoreErrors: true }
            );
            for (const f of tail) {
              sanitizeFrameTimestamp(f, () => {
                const ts = syntheticVideoUs;
                syntheticVideoUs += videoFrameStepUs;
                return ts;
              }, videoTimeBase);
              try {
                const vf = bridge.laFrameToVideoFrame(f, { timeBase: [1, 1e6] });
                frameQueue.push(vf);
                void drain();
              } catch {
              }
            }
          } catch {
          }
        }
        await drain();
        if (audioDec) {
          await decodeAudioBatch([], true);
        }
      }
    });
    throwIfAborted2();
    throwIfDrainError2();
    videoSource?.close();
    audioSource?.close();
    await output.finalize();
    if (!bufferTarget.buffer) {
      throw new Error("transcode: mediabunny produced no output buffer");
    }
    const mimeType = mimeForFormat2(outputFormat);
    const blob = new Blob([bufferTarget.buffer], { type: mimeType });
    options.onProgress?.({ percent: 100, bytesWritten: blob.size });
    return {
      blob,
      mimeType,
      container: outputFormat,
      videoCodec: videoSource ? videoCodec : void 0,
      audioCodec: audioSource ? audioCodec : void 0,
      duration: ctx.duration,
      filename: generateFilename2(ctx.name, outputFormat)
    };
    async function decodeAudioBatch(pkts, flush) {
      if (!audioDec || !audioSource) return;
      const libav = demux.libav;
      let frames;
      try {
        frames = await libav.ff_decode_multi(
          audioDec.c,
          audioDec.pkt,
          audioDec.frame,
          pkts,
          flush ? { fin: true, ignoreErrors: true } : { ignoreErrors: true }
        );
      } catch (err) {
        throw new AvbridgeError(
          ERR_TRANSCODE_DECODE,
          `transcode: audio decode failed: ${err.message}`,
          void 0
        );
      }
      for (const f of frames) {
        sanitizeFrameTimestamp(f, () => {
          const ts = syntheticAudioUs;
          const samples = f.nb_samples ?? 1024;
          const sampleRate = f.sample_rate ?? 44100;
          syntheticAudioUs += Math.round(samples * 1e6 / sampleRate);
          return ts;
        }, audioTimeBase);
        const pcm = libavFrameToInterleavedFloat32(f);
        if (!pcm) continue;
        const sample = new mb.AudioSample({
          data: pcm.data,
          format: "f32",
          numberOfChannels: pcm.channels,
          sampleRate: pcm.sampleRate,
          timestamp: (f.pts ?? 0) / 1e6
        });
        await audioSource.add(sample);
      }
    }
  } finally {
    try {
      await demux.destroy();
    } catch {
    }
  }
}
async function loadBridge() {
  try {
    const wrapper = await import('./libav-import-6MGLCXVQ.js');
    return wrapper.libavBridge;
  } catch (err) {
    throw new Error(`failed to load libavjs-webcodecs-bridge: ${err.message}`);
  }
}
function avbridgeVideoToMediabunny(c) {
  switch (c) {
    case "h264":
      return "avc";
    case "h265":
      return "hevc";
    case "vp9":
      return "vp9";
    case "av1":
      return "av1";
  }
}
function avbridgeAudioToMediabunny(c) {
  switch (c) {
    case "aac":
      return "aac";
    case "opus":
      return "opus";
    case "flac":
      return "flac";
  }
}
function qualityToMediabunny(mb, quality, override) {
  if (override !== void 0) return override;
  switch (quality) {
    case "low":
      return mb.QUALITY_LOW;
    case "medium":
      return mb.QUALITY_MEDIUM;
    case "high":
      return mb.QUALITY_HIGH;
    case "very-high":
      return mb.QUALITY_VERY_HIGH;
  }
}

// src/convert/transcode.ts
var MEDIABUNNY_CONTAINERS = /* @__PURE__ */ new Set([
  "mp4",
  "mov",
  "mkv",
  "webm",
  "ogg",
  "wav",
  "mp3",
  "flac",
  "adts"
]);
async function transcode(source, options = {}) {
  const outputFormat = options.outputFormat ?? "mp4";
  const videoCodec = options.videoCodec ?? defaultVideoCodec(outputFormat);
  const audioCodec = options.audioCodec ?? defaultAudioCodec(outputFormat);
  const quality = options.quality ?? "medium";
  validateCodecCompatibility(outputFormat, videoCodec, audioCodec);
  options.signal?.throwIfAborted();
  const ctx = await probe(source);
  options.signal?.throwIfAborted();
  if (isLibavTranscodeContainer(ctx.container)) {
    return transcodeViaLibav(ctx, options);
  }
  if (!MEDIABUNNY_CONTAINERS.has(ctx.container)) {
    throw new AvbridgeError(
      ERR_CONTAINER_NOT_SUPPORTED,
      `Cannot transcode "${ctx.container}" sources. transcode() supports mediabunny-readable containers (MP4, MKV, WebM, OGG, MP3, FLAC, WAV, MOV) and legacy containers via the libav path (AVI, ASF, FLV).`,
      `If this is a legacy container we don't yet support, use createPlayer() to play it. Transcode support for more containers is on the roadmap.`
    );
  }
  return doTranscode(ctx, outputFormat, videoCodec, audioCodec, quality, options);
}
async function attemptTranscode(ctx, outputFormat, videoCodec, audioCodec, quality, options) {
  const mb = await import('mediabunny');
  const input = new mb.Input({
    source: await buildMediabunnySourceFromInput(mb, ctx.source),
    formats: mb.ALL_FORMATS
  });
  let bytesWritten = 0;
  const useStream = !!options.outputStream;
  const bufferTarget = useStream ? null : new mb.BufferTarget();
  const streamTarget = useStream ? new mb.StreamTarget(new WritableStream({
    write(chunk) {
      bytesWritten += chunk.data.byteLength;
      const writer = options.outputStream.getWriter();
      return writer.write(chunk.data).then(() => writer.releaseLock());
    }
  })) : null;
  const output = new mb.Output({
    format: createOutputFormat(mb, outputFormat),
    target: streamTarget ?? bufferTarget
  });
  const videoOptions = options.dropVideo ? { discard: true } : {
    codec: avbridgeVideoToMediabunny2(videoCodec),
    bitrate: options.videoBitrate ?? qualityToMediabunny2(mb, quality),
    forceTranscode: true,
    ...options.width !== void 0 ? { width: options.width } : {},
    ...options.height !== void 0 ? { height: options.height } : {},
    ...options.width !== void 0 && options.height !== void 0 ? { fit: "contain" } : {},
    ...options.frameRate !== void 0 ? { frameRate: options.frameRate } : {},
    ...options.hardwareAcceleration !== void 0 ? { hardwareAcceleration: options.hardwareAcceleration } : {}
  };
  const audioOptions = options.dropAudio ? { discard: true } : {
    codec: avbridgeAudioToMediabunny2(audioCodec),
    bitrate: options.audioBitrate ?? qualityToMediabunny2(mb, quality),
    forceTranscode: true
  };
  const conversion = await mb.Conversion.init({
    input,
    output,
    video: videoOptions,
    audio: audioOptions,
    showWarnings: false
  });
  if (!conversion.isValid) {
    const reasons = conversion.discardedTracks.map((d) => `${d.track.type} track discarded: ${d.reason}`).join("; ");
    throw new Error(
      `Cannot transcode: mediabunny rejected the conversion. ${reasons || "(no reason given)"}`
    );
  }
  if (options.onProgress) {
    const onProgress = options.onProgress;
    conversion.onProgress = (p) => {
      onProgress({ percent: p * 100, bytesWritten: 0 });
    };
  }
  let abortHandler;
  if (options.signal) {
    options.signal.throwIfAborted();
    abortHandler = () => void conversion.cancel();
    options.signal.addEventListener("abort", abortHandler, { once: true });
  }
  try {
    await conversion.execute();
  } finally {
    if (abortHandler && options.signal) {
      options.signal.removeEventListener("abort", abortHandler);
    }
  }
  if (useStream) {
    return null;
  }
  if (!bufferTarget.buffer) {
    throw new Error("Transcode failed: mediabunny produced no output buffer.");
  }
  return bufferTarget.buffer;
}
function isLikelyEncoderInitError(err) {
  if (!err) return false;
  const msg = err instanceof Error ? err.message : String(err);
  const lower = msg.toLowerCase();
  return lower.includes("encoding error") || lower.includes("encoder") || lower.includes("encode failed");
}
function describeError(err) {
  if (!err) return "(unknown)";
  if (err instanceof Error) return err.message;
  return String(err);
}
var MAX_ENCODER_RETRIES = 2;
async function doTranscode(ctx, outputFormat, videoCodec, audioCodec, quality, options) {
  const notes = [];
  let buffer = null;
  let lastError;
  for (let attempt = 0; attempt <= MAX_ENCODER_RETRIES; attempt++) {
    try {
      buffer = await attemptTranscode(ctx, outputFormat, videoCodec, audioCodec, quality, options);
      if (attempt > 0) {
        notes.push(
          `Encoder failed ${attempt} time${attempt === 1 ? "" : "s"} before succeeding (known headless Chromium WebCodecs encoder init issue): ${describeError(lastError)}`
        );
      }
      break;
    } catch (err) {
      lastError = err;
      if (options.signal?.aborted) throw err;
      if (!isLikelyEncoderInitError(err)) throw err;
      if (attempt === MAX_ENCODER_RETRIES) throw err;
      await new Promise((r) => setTimeout(r, 50 * (attempt + 1)));
    }
  }
  const mimeType = mimeForFormat(outputFormat);
  const filename = generateFilename(ctx.name, outputFormat);
  if (options.outputStream) {
    options.onProgress?.({ percent: 100, bytesWritten: 0 });
    return {
      blob: new Blob([], { type: mimeType }),
      mimeType,
      container: outputFormat,
      videoCodec: options.dropVideo ? void 0 : videoCodec,
      audioCodec: options.dropAudio ? void 0 : audioCodec,
      duration: ctx.duration,
      filename,
      ...notes.length > 0 ? { notes } : {}
    };
  }
  if (!buffer) {
    throw new Error("Transcode failed: no buffer produced (this should be unreachable).");
  }
  const blob = new Blob([buffer], { type: mimeType });
  options.onProgress?.({ percent: 100, bytesWritten: blob.size });
  return {
    blob,
    mimeType,
    container: outputFormat,
    videoCodec: options.dropVideo ? void 0 : videoCodec,
    audioCodec: options.dropAudio ? void 0 : audioCodec,
    duration: ctx.duration,
    filename,
    ...notes.length > 0 ? { notes } : {}
  };
}
function defaultVideoCodec(format) {
  switch (format) {
    case "webm":
      return "vp9";
    case "mp4":
    case "mkv":
    default:
      return "h264";
  }
}
function defaultAudioCodec(format) {
  switch (format) {
    case "webm":
      return "opus";
    case "mp4":
    case "mkv":
    default:
      return "aac";
  }
}
function validateCodecCompatibility(format, videoCodec, audioCodec) {
  if (format === "webm") {
    if (videoCodec !== "vp9" && videoCodec !== "av1") {
      throw new Error(
        `WebM does not support video codec "${videoCodec}". Use "vp9" or "av1", or change outputFormat to "mp4" or "mkv".`
      );
    }
    if (audioCodec !== "opus") {
      throw new Error(
        `WebM does not support audio codec "${audioCodec}". Use "opus", or change outputFormat to "mp4" or "mkv".`
      );
    }
  }
}
function avbridgeVideoToMediabunny2(c) {
  switch (c) {
    case "h264":
      return "avc";
    case "h265":
      return "hevc";
    case "vp9":
      return "vp9";
    case "av1":
      return "av1";
  }
}
function avbridgeAudioToMediabunny2(c) {
  switch (c) {
    case "aac":
      return "aac";
    case "opus":
      return "opus";
    case "flac":
      return "flac";
  }
}
function qualityToMediabunny2(mb, quality) {
  switch (quality) {
    case "low":
      return mb.QUALITY_LOW;
    case "medium":
      return mb.QUALITY_MEDIUM;
    case "high":
      return mb.QUALITY_HIGH;
    case "very-high":
      return mb.QUALITY_VERY_HIGH;
  }
}

export { transcode };
